{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:70% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:70% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import seaborn as sns\n",
    "import scipy as sp\n",
    "from scipy.signal import savgol_filter as savgol \n",
    "from copy import deepcopy    \n",
    "from random import choices\n",
    "from scipy.stats import ttest_ind\n",
    "from sklearn.metrics import auc\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "from matplotlib import colors as mcolors\n",
    "from matplotlib import cm\n",
    "from matplotlib.patches import Ellipse\n",
    "import matplotlib.transforms as transforms\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "from mpl_toolkits.axes_grid1.inset_locator import zoomed_inset_axes, mark_inset\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "import matplotlib\n",
    "npg_colors = ['#00A087FF','#E64B35FF','#3C5488FF','#4DBBD5FF','#F39B7FFF','#8491B4FF','#91D1C2FF','#DC0000FF','#7E6148FF','#B09C85FF']\n",
    "npg_color_list = []\n",
    "for this_color in npg_colors: \n",
    "    npg_color_list.append(list(int(this_color[1:][i:i+2], 16)/255 for i in (0, 2, 4)))\n",
    "# plt.style.use('dark_background')\n",
    "# dark_colors = plt.rcParams['axes.prop_cycle'].by_key()['color']  \n",
    "plt.rcParams.update({'font.size':25});\n",
    "plt.rcParams.update({'axes.linewidth':1.5})\n",
    "plt.rc('lines',linewidth=3);\n",
    "plt.rcParams['axes.spines.right'] = False\n",
    "plt.rcParams['axes.spines.top'] = False \n",
    "ncolors = 10\n",
    "plt.rc('axes', axisbelow=True)\n",
    "\n",
    "from preprocess import *\n",
    "from dmd import *\n",
    "from sensor_placement import *\n",
    "# from query_uniprot import *\n",
    "from regions import *\n",
    "from plot_ellipsoid import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Import data and metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# normalized gene expression in TPM generated with Kallisto\n",
    "datadir = 'data/tpm_removed_low_count_genes.csv' \n",
    "df = pd.read_csv(datadir)\n",
    "\n",
    "ntimepts = 12 # total number of timepoints in df\n",
    "reps = [0,2] # downselect to the replicates used in our work (2nd replicate not used due to experimental issue)\n",
    "nreps = 3 # total number of replicates in df\n",
    "# tp_list = list(range(2,ntimepts-1)) # the timepoints to be considered for modeling \n",
    "# malathion was introduced AFTER the 2nd timepoint not using the last timepoint due to experimental issue\n",
    "\n",
    "# getting sample labels and transcript IDs from the df\n",
    "sampleLabels = list(df.columns[1:])\n",
    "txIDs = list(df.iloc[:,0])\n",
    "# getting transcriptome dynamics\n",
    "df = df.iloc[:,1:] "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Preprocess to get several different subsets of the data, metadata (gene names, reps, etc.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "num genes x num timepoints x num replicates\n",
      "(624, 9, 2)\n"
     ]
    }
   ],
   "source": [
    "data_c_orig, data_t_orig, data_c, data_t, keepers, data_c_keep, data_t_keep, \\\n",
    "                                txIDs_keep, data_fc, data_fc_norm = \\\n",
    "                                process_df(df,sampleLabels,nreps,reps,ntimepts,txIDs)\n",
    "\n",
    "print('num genes x num timepoints x num replicates')\n",
    "print(data_fc_norm.shape) # genes x timepoints x replicates\n",
    "\n",
    "genbank_path = 'data/GCA_000009225.1_ASM922v1_genomic.gbff'\n",
    "cdsFasta_path = 'data/GCA_000009225.1_ASM922v1_cds_from_genomic.fa'\n",
    "genes_keep, locus_tags_keep, locations = getRecords(cdsFasta_path,genbank_path,txIDs_keep)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Perform dynamic mode decomposition on transcriptome dynamics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "r2_score for n-step prediction (reduced): 9.393e-01\n"
     ]
    }
   ],
   "source": [
    "r = 10\n",
    "A,Atilde,data_fc_norm_red,U_r,cd,eVals_r,eVecs_r,Phi,b_r0,b_r1 = \\\n",
    "        dmd(data_fc_norm,rank_reduce=True,r=r,trim=False,trimThresh=2.5e-3) #trimThesh=2.5e-3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Observability maximization with closed form solution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First, we will compute the Gram matrix which encodes all known information \n",
    "nT = 8 # used 8 in manuscript\n",
    "# this G is computed using non-projected data (can be slow for high-dimensional data)\n",
    "# A = (np.absolute(A) > 2.5e-3) * A\n",
    "G = gram_matrix(A,data_fc_norm[:,0,:],nT=nT,reduced=False)\n",
    "# Gt is computed using projected data and KO and Gt_recon is simply projected back up to original dimension\n",
    "Gt, Gt_recon = gram_matrix(Atilde,data_fc_norm_red[:,0,:],nT=100,reduced=True,projection_matrix=U_r)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute eigendecomposition of the Gram matrices\n",
    "# right eigenvectors of G (columns of V) are rows of the gene sampling matrix \n",
    "D, V = sp.linalg.eig(G,left=False,right=True)\n",
    "Dt,Vt = sp.linalg.eig(Gt,left=False,right=True)\n",
    "Dt_recon, Vt_recon = sp.linalg.eig(Gt_recon,left=False,right=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get gene sampling weights\n",
    "# W = Vt_recon[:,0:p].T\n",
    "W = V[:,0:1].T\n",
    "# W = V[:,1:2].T\n",
    "sorted_inds = list(np.argsort(np.abs(W[0,:]))) # based on top eigenvector only\n",
    "sorted_tags = [locus_tags_keep[x] for x in sorted_inds]\n",
    "Wsorted = deepcopy(W)[:,sorted_inds] # Used to generate percent rank of genes\n",
    "\n",
    "# normalize gene sampling weights\n",
    "gene_std = np.mean(np.std(data_fc_norm,axis=2),axis=1)\n",
    "Wn = W[0]/( gene_std )\n",
    "sorted_inds_n = list(np.argsort(np.abs(Wn))) \n",
    "sorted_tags_n = [locus_tags_keep[x] for x in sorted_inds_n]\n",
    "Wsorted_n = deepcopy(Wn)[sorted_inds_n]\n",
    "\n",
    "# apply shrinkage to standard deviations based on sampling weights, then normalize\n",
    "z, cov = np.polyfit(x=W[0], y=gene_std, deg=2, cov=True)\n",
    "x_w = np.linspace(W[0].min(),W[0].max(),len(W[0]))\n",
    "std_x_w = z[0]*x_w**2+z[1]*x_w+z[2]\n",
    "Ws = W[0] / std_x_w # normalized via shrinkage estimated std\n",
    "sorted_inds_s = list(np.argsort(np.abs(Ws))) \n",
    "sorted_tags_s = [locus_tags_keep[x] for x in sorted_inds_s]\n",
    "Wsorted_s = deepcopy(Ws)[sorted_inds_s]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Observability maximization with SQP (force weights to be > 0 and enforce sparsity) \n",
    "- Takes between 1 and 10 minutes per run for $O(10^3)$ genes (depends on initialization)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# doRun = False\n",
    "# doSave = False\n",
    "# if doRun: \n",
    "#     Clist = []\n",
    "#     for ii in range(10): # run the observability analysis 10 times\n",
    "#         C,seed = energy_maximization_with_uni_dist(data_fc_norm,A,data_fc_norm.shape[1],data_fc_norm.shape[1]-1,IC=0) \n",
    "#         C = (C/C.max()).T # normalizing C to be b/w 0 and 1\n",
    "#         C = list(np.squeeze(C))\n",
    "#         Clist.append(C)\n",
    "\n",
    "# if doSave: \n",
    "# # dump observability result \n",
    "#     pickle.dump(Clist,open('data/Clist.pkl','wb'))\n",
    "        \n",
    "        \n",
    "# if not doRun: # grab from previously saved pickle \n",
    "#     # retrieve Clist from pickle file\n",
    "#     Clist = pickle.load(open('data/Clist.pkl','rb'))\n",
    "    \n",
    "# # sort from lowest to highest sampling weight and get mean/std over the runs\n",
    "# C = list(np.mean(Clist,axis=0))\n",
    "# Csorted = np.sort(C)\n",
    "# CsortedInds = list(np.argsort(C)) # from lowest to highest rank\n",
    "\n",
    "# # to get the error across runs, need each list in Clist to be sorted as in CsortedInds\n",
    "# Ctmp = []\n",
    "# for ii in range(len(Clist)):\n",
    "#     Ctmp.append(list(np.array(Clist[ii])[CsortedInds]))\n",
    "    \n",
    "# Csorted_err = np.std(np.array(Ctmp),axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Measure reconstruction accuracy for varying timepoints and number of genes selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# takes several minutes to run\n",
    "\n",
    "samplingFreq = 30\n",
    "\n",
    "doRun = False\n",
    "if doRun:\n",
    "    rho_T10 = reconstruct_x0_sequential(data_fc_norm,10,A,Wn.reshape(-1,len(Wn)),sorted_inds_n,samplingFreq=samplingFreq,order='top') # second argument is number of timepoints\n",
    "    rho_T9 = reconstruct_x0_sequential(data_fc_norm,9,A,Wn.reshape(-1,len(Wn)),sorted_inds_n,samplingFreq=samplingFreq,order='top')\n",
    "    rho_T8 = reconstruct_x0_sequential(data_fc_norm,8,A,Wn.reshape(-1,len(Wn)),sorted_inds_n,samplingFreq=samplingFreq,order='top')\n",
    "    rho_T7 = reconstruct_x0_sequential(data_fc_norm,7,A,Wn.reshape(-1,len(Wn)),sorted_inds_n,samplingFreq=samplingFreq,order='top')\n",
    "    rho_T6 = reconstruct_x0_sequential(data_fc_norm,6,A,Wn.reshape(-1,len(Wn)),sorted_inds_n,samplingFreq=samplingFreq,order='top')\n",
    "    rho_T5 = reconstruct_x0_sequential(data_fc_norm,5,A,Wn.reshape(-1,len(Wn)),sorted_inds_n,samplingFreq=samplingFreq,order='top')\n",
    "    rho_T4 = reconstruct_x0_sequential(data_fc_norm,4,A,Wn.reshape(-1,len(Wn)),sorted_inds_n,samplingFreq=samplingFreq,order='top')\n",
    "    rho_T3 = reconstruct_x0_sequential(data_fc_norm,3,A,Wn.reshape(-1,len(Wn)),sorted_inds_n,samplingFreq=samplingFreq,order='top')\n",
    "    rho_T2 = reconstruct_x0_sequential(data_fc_norm,2,A,Wn.reshape(-1,len(Wn)),sorted_inds_n,samplingFreq=samplingFreq,order='top')\n",
    "    rho_T1 = reconstruct_x0_sequential(data_fc_norm,1,A,Wn.reshape(-1,len(Wn)),sorted_inds_n,samplingFreq=samplingFreq,order='top')\n",
    "    rho_list = [rho_T10,rho_T9,rho_T8,rho_T7,rho_T6,rho_T5,rho_T4,rho_T3,rho_T2,rho_T1]\n",
    "    \n",
    "doRunRand = False\n",
    "if doRunRand: \n",
    "    rho_T10_rand = reconstruct_x0_sequential(data_fc_norm,10,A,W,sorted_inds,samplingFreq=samplingFreq,order='random')\n",
    "    rho_T9_rand = reconstruct_x0_sequential(data_fc_norm,9,A,W,sorted_inds,samplingFreq=samplingFreq,order='random')\n",
    "    rho_T8_rand = reconstruct_x0_sequential(data_fc_norm,8,A,W,sorted_inds,samplingFreq=samplingFreq,order='random')\n",
    "    rho_T7_rand = reconstruct_x0_sequential(data_fc_norm,7,A,W,sorted_inds,samplingFreq=samplingFreq,order='random')\n",
    "    rho_T6_rand = reconstruct_x0_sequential(data_fc_norm,6,A,W,sorted_inds,samplingFreq=samplingFreq,order='random')\n",
    "    rho_T5_rand = reconstruct_x0_sequential(data_fc_norm,5,A,W,sorted_inds,samplingFreq=samplingFreq,order='random')\n",
    "    rho_T4_rand = reconstruct_x0_sequential(data_fc_norm,4,A,W,sorted_inds,samplingFreq=samplingFreq,order='random')\n",
    "    rho_T3_rand = reconstruct_x0_sequential(data_fc_norm,3,A,W,sorted_inds,samplingFreq=samplingFreq,order='random')\n",
    "    rho_T2_rand = reconstruct_x0_sequential(data_fc_norm,2,A,W,sorted_inds,samplingFreq=samplingFreq,order='random')\n",
    "    rho_T1_rand = reconstruct_x0_sequential(data_fc_norm,1,A,W,sorted_inds,samplingFreq=samplingFreq,order='random')\n",
    "\n",
    "doSave = False\n",
    "if doSave: \n",
    "    pickle.dump([rho_T10,rho_T9,rho_T8,rho_T7,rho_T6,rho_T5,rho_T4,rho_T3,rho_T2,rho_T1],open('data/reconstructions.pkl','wb'))\n",
    "#     pickle.dump([rho_T10_rand,rho_T9_rand,rho_T8_rand,rho_T7_rand,rho_T6_rand,rho_T5_rand,rho_T4_rand,rho_T3_rand,rho_T2_rand,rho_T1_rand],open('data/reconstructions_random.pkl','wb'))\n",
    "\n",
    "if not doRun:\n",
    "    rho_list = pickle.load(open('data/reconstructions.pkl','rb'))\n",
    "if not doRunRand:\n",
    "    rho_rand_list = pickle.load(open('data/reconstructions_random.pkl','rb'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Results dataframe with selected encoder genes and metadata\n",
    "The uniprot service from the bioservices package is failing (https://bioservices.readthedocs.io/en/main/), going to manually define the gene names, locus_tags, and proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# # their locus_tags\n",
    "# tags = ['6124','0841','1823','5502','1836','4736','3761','4150','0376','1815','0953','1358','1302A','4612','1380']\n",
    "# tags = ['PFLU_'+tag for tag in tags]\n",
    "# search_names = 'genes,protein names,comment(FUNCTION),go,go(biological process),go(molecular function)'\n",
    "\n",
    "# df_up_new = process_query(tags,search_names,locus_tags_keep,W[0],Wsorted[0])\n",
    "# # adjust long gene names for manuscript\n",
    "# df_up_new.gene[list(df_up_new.locus_tag).index('PFLU_3761')] = 'Uncharacterized \\n protein I'\n",
    "# df_up_new.gene[list(df_up_new.locus_tag).index('PFLU_1358')] = 'Uncharacterized \\n protein II'\n",
    "# df_up_new.gene[list(df_up_new.locus_tag).index('PFLU_4612')] = 'Putative outer membrane porin \\n A protein'\n",
    "# df_up_new.gene[list(df_up_new.locus_tag).index('PFLU_0376')] = 'Putative ABC transport \\n system'\n",
    "# df_up_new.gene[list(df_up_new.locus_tag).index('PFLU_4736')] = 'anti-sigma \\n 28 factor'\n",
    "\n",
    "# doSave = False\n",
    "# if doSave: \n",
    "#     df_up_new.to_csv('data/encoder_genes_metadata.csv',sep=',',index=False)\n",
    "# df_up_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "tags = ['6124','0841','1823','5502','1836','4736','3761','4150','0376','1815','0953','1358','1302A','4612','1380']\n",
    "tags = ['PFLU_'+tag for tag in tags]\n",
    "df_up_new = pd.DataFrame([],columns = ['gene','locus_tag'])\n",
    "df_up_new['locus_tag'] = tags\n",
    "df_up_new['gene'] = ['atpB','petA','sucC','rpoA','fabA','anti-sigma \\n 28 factor', \n",
    "                     'Uncharacterized \\n protein I','cspA2','Putative ABC transport \\n system',\n",
    "                    'gltA','lpxC','Uncharacterized \\n protein II','capB',\n",
    "                     'Putative outer membrane porin \\n A protein','acrA']\n",
    "\n",
    "# get observability rank in percentage from the sampling weights\n",
    "tag_inds = []\n",
    "for tag in tags:\n",
    "    ind = locus_tags_keep.index(tag)\n",
    "    tag_inds.append(ind)\n",
    "df_up_new['tag_ind'] = tag_inds\n",
    "\n",
    "rank_tag_inds = []\n",
    "for ii in tag_inds:\n",
    "    rank_tag_inds.append(len(W[0]) - list(Wsorted[0]).index(W[0][ii]))     \n",
    "    rank_tag_inds_per = [100-np.round((rank_ind-1) / len(W[0]) * 100,2) for rank_ind in rank_tag_inds]\n",
    "df_up_new['obs_rank'] = rank_tag_inds_per\n",
    "\n",
    "rank_tag_inds = []\n",
    "for ii in tag_inds:\n",
    "    rank_tag_inds.append(len(Wn) - list(Wsorted_n).index(Wn[ii]))     \n",
    "    rank_tag_inds_per = [100-np.round((rank_ind-1) / len(Wn) * 100,2) for rank_ind in rank_tag_inds]\n",
    "df_up_new['obs_stand_rank'] = rank_tag_inds_per\n",
    "\n",
    "rank_tag_inds  = []\n",
    "for ii in tag_inds:\n",
    "    rank_tag_inds.append(len(Ws) - list(Wsorted_s).index(Ws[ii]))     \n",
    "    rank_tag_inds_per = [100-np.round((rank_ind-1) / len(Ws) * 100,2) for rank_ind in rank_tag_inds]\n",
    "df_up_new['obs_stand_shrink_rank'] = rank_tag_inds_per\n",
    "\n",
    "df_up_new"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get promoter sequences corresponding to selected encoder genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "intergenic_records = get_interregions(genbank_path,intergene_length=100)\n",
    "# get promoter sequences (intergenic regions) for all the genes modeled\n",
    "promoter_records = get_promoterregions_v2(locations,locus_tags_keep,intergenic_records,genbank_path,rel_dist_thresh=5000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figures\n",
    "- Figure 2 and 3 are generated here, for Figures 4 and 5 see reporters.ipynb and virtual_sensors.ipynb, respectively"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "figDir = '/Users/aqib/Desktop/BCCL/koopman-sensor-placement/NC_submission/figures/'\n",
    "corr_tspan = list(range(10,100,10)) # for xaxis in many figures"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### How does connectedness of the dynamics affect the observability maximization problem? Number of timepoints for observability maximization?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# savefig = False\n",
    "\n",
    "# s_thresh_list = [0.0, 0.006, 0.01] \n",
    "# colors = ['tab:blue','tab:orange','tab:green']\n",
    "# fig, ax = plt.subplots(1,2,figsize=(12,5))\n",
    "# for ii, s_thresh in enumerate(s_thresh_list):\n",
    "#     As = (np.abs(A) > s_thresh) * A\n",
    "#     Gs = gram_matrix(As,data_fc_norm[:,0,:],nT=8,reduced=False)\n",
    "#     _, Vs = sp.linalg.eig(Gs,left=False,right=True)\n",
    "#     Vs = Vs[:,0:1].T[0]\n",
    "#     Vs = (Vs - np.min(Vs)) / (np.max(Vs) - np.min(Vs))\n",
    "#     ax[0].plot(np.sort(Vs),label=f' K > {s_thresh}')\n",
    "#     sns.histplot(Vs,ax=ax[1],color=colors[ii],alpha=0.7)\n",
    "# ax[0].legend(frameon=False,fontsize=22)\n",
    "# ax[0].set_xlabel('Gene'); ax[0].set_ylabel('Sampling weights',fontsize=27)\n",
    "# ax[1].set_xlabel('Sampling weights'); ax[1].set_ylabel('Genes',fontsize=27)\n",
    "# plt.tight_layout()\n",
    "# if savefig: \n",
    "#     plt.savefig(figDir + 'supplementalFigs/vary_sparse_thresh_for_obs_max.pdf',bbox_inches='tight',transparent=True,dpi=300)\n",
    "# plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Subfigures for FIgure 2 in manuscript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# plot the eigenvalues on the unit circle\n",
    "\n",
    "savefig = False\n",
    "# automate mode pairing (data is real-valued, complex eigs comes with a conjugate pair -- no need to plot both)\n",
    "mode_pairs = []\n",
    "cnt = 0\n",
    "while cnt in range(len(eVals_r)): \n",
    "    if cnt+1 != len(eVals_r): # have we reached the boundary?\n",
    "        if np.abs(eVals_r[cnt]) == np.abs(eVals_r[cnt+1]): # if no, then check if the two are c.c. pairs\n",
    "            mode_pairs.append([cnt,cnt+1])\n",
    "            cnt += 2\n",
    "        else: # if no, and not c.c. pairs\n",
    "            mode_pairs.append([cnt])\n",
    "            cnt += 1\n",
    "    else: # if boundary\n",
    "        mode_pairs.append([cnt])\n",
    "        cnt += 1\n",
    "print('mode pairs: ', mode_pairs)\n",
    "\n",
    "# size the eigenvalue marker by corresponding mode power ||Phi_i||\n",
    "mode_powers = np.linalg.norm(Phi,axis=0)\n",
    "my_ms = [9*2**n for n in mode_powers] # the 9*2**n is an arbitrary scaling for marker size\n",
    "\n",
    "fig1,ax1 = plt.subplots(1,1,figsize=(6,6))\n",
    "fig2,ax2 = plt.subplots(1,1,figsize=(6,6))\n",
    "theta = np.linspace(0,2*np.pi,200)\n",
    "ax1.plot(np.cos(theta),np.sin(theta),c='black',lw=1.5)\n",
    "\n",
    "markers = ['o','s','d','^','v','>']\n",
    "cmap = matplotlib.cm.get_cmap('Greys')\n",
    "grad_colors = [list(cmap(x)[0:3]) for x in np.linspace(0.1,0.9,6)]\n",
    "for ii in range(len(mode_pairs)):\n",
    "    scaled_mode = np.real(eVals_r[ii]*b_r1[mode_pairs[ii][0]])\n",
    "\n",
    "    ax2.plot(corr_tspan,savgol(scaled_mode,3,1),'-',c=npg_colors[ii],lw=4,label='Mode ' + str(mode_pairs[ii]))\n",
    "#     ax1.plot(np.real(eVals_r[mode_pairs[ii][0]]),np.imag(eVals_r[mode_pairs[ii][0]]),\n",
    "#                 marker=markers[ii],mfc=npg_color_list[ii]+[0.8],ms=my_ms[mode_pairs[ii][0]],mec='k',mew=2)\n",
    "#     if len(mode_pairs[ii]) == 2: \n",
    "#         ax1.plot(np.real(eVals_r[mode_pairs[ii][1]]),np.imag(eVals_r[mode_pairs[ii][1]]),\n",
    "#                 marker=markers[ii],mfc=npg_color_list[ii]+[0.8],ms=my_ms[mode_pairs[ii][0]],mec='k',mew=2)\n",
    "    ax1.plot(np.real(eVals_r[mode_pairs[ii][0]]),np.imag(eVals_r[mode_pairs[ii][0]]),\n",
    "                marker=markers[ii],mfc=grad_colors[ii],ms=my_ms[mode_pairs[ii][0]],mec='k',mew=1.5)\n",
    "    if len(mode_pairs[ii]) == 2: \n",
    "        ax1.plot(np.real(eVals_r[mode_pairs[ii][1]]),np.imag(eVals_r[mode_pairs[ii][1]]),\n",
    "                marker=markers[ii],mfc=grad_colors[ii],ms=my_ms[mode_pairs[ii][0]],mec='k',mew=1.5)\n",
    "\n",
    "ax1.axis('equal')\n",
    "ax1.set_xlabel(r'$\\mathregular{Re(\\lambda)}$',fontsize=27)\n",
    "ax1.set_ylabel(r'$\\mathregular{Im(\\lambda)}$',fontsize=29)\n",
    "ax1.set_ylim(-1,1)\n",
    "ax1.set_yticks(np.arange(-1,1.5,0.5))\n",
    "ax1.xaxis.set_tick_params(which='both', size=6, width=1.5, direction='out')\n",
    "ax1.yaxis.set_tick_params(which='both', size=6, width=1.5, direction='out')\n",
    "\n",
    "ax2.set_xlabel('Time (minutes)',fontsize=22)\n",
    "ax2.set_ylabel('Dynamic mode amplitude',fontsize=22)\n",
    "ax2.legend(frameon=False,ncol=1)\n",
    "ax2.xaxis.set_tick_params(which='both', size=5, width=2, direction='in')\n",
    "ax2.yaxis.set_tick_params(which='both', size=5, width=2, direction='in')\n",
    "\n",
    "for side in ax1.spines.keys():\n",
    "    ax2.spines[side].set_linewidth(2)\n",
    "\n",
    "ax2.grid(linestyle=':',linewidth=1.75,dashes=(2,4))\n",
    "    \n",
    "if savefig: \n",
    "    fig1.savefig(figDir + 'fig2/eigvals.pdf',bbox_inches='tight',transparent=True,dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# make a plot that shows magnitude of eigvals for varying r\n",
    "# supplemental figure\n",
    "\n",
    "savefig = False\n",
    "\n",
    "plt.rcParams['axes.spines.top'] = False\n",
    "plt.rcParams['axes.spines.right'] = False\n",
    "\n",
    "r_list = [6,8,10,12,14,16]\n",
    "evals_list,cd_list = [],[]\n",
    "for rr in r_list: \n",
    "    _,_,_,_,cdr,evals,_,_,_,_ = \\\n",
    "                            dmd(data_fc_norm,rank_reduce=True,r=rr,trim=True,trimThresh=2e-3)\n",
    "    evals_list.append(evals)\n",
    "    cd_list.append(cdr)\n",
    "\n",
    "# fig,ax = plt.subplots(1,1,figsize=(7,7))\n",
    "# my_ms = ['>','o','<','v','^']\n",
    "# for ii,evals in enumerate(evals_list):\n",
    "#     ax.plot(np.sort(np.abs(evals))[::-1],my_ms[ii],mec='k',c=npg_colors[ii],mew=1,ms=12\n",
    "#             ,label='r = '+str(r_list[ii]))\n",
    "# for side in ax.spines.keys():\n",
    "#     ax.spines[side].set_linewidth(2.5)\n",
    "# ax.set_ylabel(r'$|\\mathregular{\\lambda}|$',fontsize=26)\n",
    "# ax.set_xticks(range(0,20,5))\n",
    "# ax.xaxis.set_tick_params(which='both', size=5, width=2, direction='in')\n",
    "# ax.yaxis.set_tick_params(which='both', size=5, width=2, direction='in')\n",
    "# ax.grid(linestyle=':',linewidth=1.75,dashes=(2,4))\n",
    "# fig.legend(frameon=False,bbox_to_anchor=(1,0.9))\n",
    "# if savefig: \n",
    "#     plt.savefig(figDir + 'fig2/magEvals_vary_numModes.pdf',bbox_inches='tight',transparent=True,dpi=300)\n",
    "# plt.show()\n",
    "\n",
    "fig,axs = plt.subplots(2,3,figsize=(15,10))\n",
    "theta = np.linspace(0,2*np.pi,100)\n",
    "my_ms = ['o','>','<','v','^','s']\n",
    "for ii,ax in enumerate(axs.reshape(-1)):\n",
    "    ax.plot(np.cos(theta),np.sin(theta),c='k',lw=1.5)\n",
    "    ax.plot(np.real(evals_list[ii]),np.imag(evals_list[ii]),my_ms[ii],ms=12,label='r = '+str(r_list[ii]),\n",
    "            c='tab:blue',mec='k')\n",
    "    ax.set_title(r'$R^2=$' + str(np.round(cd_list[ii],2)))\n",
    "    for side in ax.spines.keys():\n",
    "        ax.spines[side].set_linewidth(2.5)\n",
    "    ax.xaxis.set_tick_params(which='both', size=5, width=2, direction='in')\n",
    "    ax.yaxis.set_tick_params(which='both', size=5, width=2, direction='in')\n",
    "    ax.axis('equal')\n",
    "#         ax.grid(linestyle=':',linewidth=1.75,dashes=(2,4))\n",
    "fig.legend(frameon=False,bbox_to_anchor=(1.025,0.663))\n",
    "fig.text(0.5, 0.05, r'$Re(\\lambda)$', ha='center',fontsize=24);\n",
    "fig.text(0.05, 0.5, r'$Im(\\lambda)$', va='center', rotation='vertical',fontsize=24);\n",
    "if savefig: \n",
    "    plt.savefig(figDir + 'supplementalFigs/evals_vary_numModes.pdf',bbox_inches='tight',transparent=True,dpi=300)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "r_list = list(range(1,17))\n",
    "cd_list = []\n",
    "for rr in r_list: \n",
    "    _,_,_,_,cdr,_,_,_,_,_ = \\\n",
    "                            dmd(data_fc_norm,rank_reduce=True,r=rr,trim=True,trimThresh=2e-3)\n",
    "    cd_list.append(cdr)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "savefig = False\n",
    "\n",
    "fig,ax = plt.subplots(1,1,figsize=(8.5,5))\n",
    "color_list = ['tab:gray']*16; color_list[9] = 'tab:blue'\n",
    "g = sns.barplot(r_list,cd_list,ax=ax,palette=color_list,saturation=1,edgecolor='k',alpha=0.75)\n",
    "# g.legend_.remove();\n",
    "ax.xaxis.set_tick_params(which='both', size=6, width=1.5, direction='out',labelsize=25)\n",
    "ax.yaxis.set_tick_params(which='both', size=6, width=1.5, direction='out',labelsize=23)\n",
    "ax.set_xticks(list(range(1,17,2)));\n",
    "ax.set_xlabel('Number of DMD modes',fontsize=27); ax.set_ylabel(r'$R^2$',fontsize=32)\n",
    "if savefig: \n",
    "    plt.savefig(figDir + 'fig2/R2_vary_numModes.pdf',bbox_inches='tight',transparent=True,dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "##### cluster genes based on modes\n",
    "\n",
    "savefig = False\n",
    "\n",
    "Phi_mag = np.abs(Phi)\n",
    "\n",
    "kmax = 20 # get most impacted genes\n",
    "kmax_inds = []\n",
    "for ii in range(Phi_mag.shape[1]):\n",
    "    kmax_inds.append(Phi_mag[:,ii].argsort()[-kmax:])\n",
    "\n",
    "heights = [0.85,1.25]\n",
    "my_figsize = (4.8,6)\n",
    "\n",
    "cmap = matplotlib.cm.get_cmap('Greys')\n",
    "grad_colors = [list(cmap(x)[0:3]) for x in np.linspace(0.1,0.9,6)]\n",
    "\n",
    "fig1,axs1 = plt.subplots(2,1,figsize=my_figsize,gridspec_kw={'height_ratios':heights},sharey='row',sharex=True)\n",
    "scaled_mode = np.real(eVals_r[0]*b_r0[0])\n",
    "axs1[0].plot(corr_tspan,savgol(scaled_mode,3,1),'o-',ms=10.5,mec='k',mew=2,c='k',mfc=grad_colors[0],lw=1.5)\n",
    "for ii in range(6,16): # only plot the top 10 \n",
    "    axs1[1].plot(corr_tspan,savgol(data_fc[kmax_inds[0][ii],:,1],3,1),'o-',c='k',mfc=grad_colors[0],\n",
    "           ms=10.5,mec='k',mew=1.5,lw=1.5)\n",
    "for jj,ax in enumerate(axs1.reshape(-1)):\n",
    "    ax.xaxis.set_tick_params(which='both', size=6, width=1.5, direction='out')\n",
    "    ax.yaxis.set_tick_params(which='both', size=6, width=1.5, direction='out')\n",
    "plt.tight_layout()    \n",
    "axs1[0].set_ylim(-24,20)\n",
    "axs1[1].set_ylim(0.5,1.55)\n",
    "axs1[0].set_ylabel('Mode 1 \\n dynamics',fontsize=25) \n",
    "axs1[1].set_ylabel('Fold change of \\n 1st mode genes',fontsize=25) \n",
    "\n",
    "fig2,axs2 = plt.subplots(2,1,figsize=my_figsize,gridspec_kw={'height_ratios':heights},sharey='row',sharex=True)\n",
    "scaled_mode = np.real(eVals_r[1]*b_r1[1])\n",
    "axs2[0].plot(corr_tspan,savgol(scaled_mode,3,1),'s-',ms=10.5,mec='k',mew=2,c='k',mfc=grad_colors[1],lw=1.5)\n",
    "for ii in range(6,16): # only plot the top 10 \n",
    "    axs2[1].plot(corr_tspan,savgol(data_fc[kmax_inds[1][ii],:,1],3,1),'s-',c='k',mfc=grad_colors[1],\n",
    "           ms=10.5,mec='k',mew=1.5,lw=1.5)\n",
    "for jj,ax in enumerate(axs2.reshape(-1)):\n",
    "    ax.xaxis.set_tick_params(which='both', size=6, width=1.5, direction='out')\n",
    "    ax.yaxis.set_tick_params(which='both', size=6, width=1.5, direction='out')\n",
    "plt.tight_layout()    \n",
    "axs2[0].set_ylim(-24,20)\n",
    "axs2[1].set_ylim(0.5,1.55)\n",
    "axs2[0].set_ylabel('Mode 2 \\n dynamics',fontsize=25) \n",
    "axs2[1].set_ylabel('Fold change of \\n 2nd mode genes',fontsize=25) \n",
    "    \n",
    "fig3,axs3 = plt.subplots(2,1,figsize=my_figsize,gridspec_kw={'height_ratios':heights},sharey='row',sharex=True)\n",
    "scaled_mode = np.real(eVals_r[5]*b_r1[5])\n",
    "axs3[0].plot(corr_tspan,savgol(scaled_mode,3,1),'^-',ms=12,mec='k',mew=2,c='k',mfc=grad_colors[3],lw=1.5)\n",
    "for ii in range(10,kmax): # only plot the top 10 \n",
    "    axs3[1].plot(corr_tspan,savgol(data_fc[kmax_inds[3][ii],:,1],3,1),'^-',c='k',mfc=grad_colors[3],\n",
    "           ms=12,mec='k',mew=1.5,lw=1.5)\n",
    "for jj,ax in enumerate(axs3.reshape(-1)):\n",
    "    ax.xaxis.set_tick_params(which='both', size=6, width=1.5, direction='out')\n",
    "    ax.yaxis.set_tick_params(which='both', size=6, width=1.5, direction='out')\n",
    "plt.tight_layout()    \n",
    "axs3[0].set_ylim(-24,20)\n",
    "axs3[1].set_ylim(0.5,1.55)\n",
    "axs3[0].set_ylabel('Mode 6 \\n dynamics',fontsize=25) \n",
    "axs3[1].set_ylabel('Fold change of \\n 6th mode genes',fontsize=25) \n",
    "\n",
    "if savefig: \n",
    "    fig1.savefig(figDir + 'fig2/mode1_genes.pdf',bbox_inches='tight',transparent=True,dpi=300)\n",
    "    fig2.savefig(figDir + 'fig2/mode2_genes.pdf',bbox_inches='tight',transparent=True,dpi=300)\n",
    "    fig3.savefig(figDir + 'fig2/mode6_genes.pdf',bbox_inches='tight',transparent=True,dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "##### visualize the predictions for several randomly selected genes\n",
    "savefig = False\n",
    "\n",
    "plt.rcParams['axes.spines.right'] = False\n",
    "plt.rcParams['axes.spines.top'] = False \n",
    "\n",
    "X_pred, cd_ = n_step_prediction(A,data_fc_norm,data_fc_norm.shape[1],data_fc_norm.shape[2])\n",
    "Xpred = X_pred.reshape(len(A),data_fc_norm.shape[1],data_fc_norm.shape[2],order='F')\n",
    "\n",
    "fig,axs = plt.subplots(3,2,figsize=(7,7),sharex=True,sharey=True)\n",
    "fig.delaxes(axs[-1,-1])\n",
    "pinds = [3,100,583,164,309] # plot these genes (selected randomly)\n",
    "for ii,ax in enumerate(axs.reshape(-1)):\n",
    "    if ii < len(pinds):\n",
    "        pind = pinds[ii]\n",
    "        act_mean = np.mean(data_fc_norm[pind],axis=1)\n",
    "        pred_mean = np.mean(Xpred[pind],axis=1)\n",
    "        act_std = np.std(data_fc_norm[pind],axis=1)\n",
    "        pred_std = np.std(Xpred[pind],axis=1)\n",
    "        ax.plot(corr_tspan,act_mean,'o-',ms=8,lw=2,c='tab:blue')\n",
    "        ax.plot(corr_tspan,pred_mean,'--',lw=4,c='tab:orange')\n",
    "        ax.fill_between(corr_tspan,act_mean-act_std,act_mean+act_std,color='tab:blue',alpha=0.2)\n",
    "        ax.fill_between(corr_tspan,pred_mean-pred_std,pred_mean+pred_std,color='tab:orange',alpha=0.2)\n",
    "        # also plot a marker indicating that the IC is given as input to the model\n",
    "        ax.plot(10,act_mean[0],'s',ms=18,c='m',mfc='None',mew=2.5)\n",
    "        ax.set_xticks(range(20,100,20))\n",
    "        ax.xaxis.set_tick_params(which='both', size=6, width=1.5, direction='out',labelsize=23)\n",
    "        ax.yaxis.set_tick_params(which='both', size=6, width=1.5, direction='out',labelsize=23)\n",
    "        if genes_keep[pind] != 'N/A':\n",
    "            ax.set_title(genes_keep[pind],fontsize=22)\n",
    "        else: \n",
    "            ax.set_title(locus_tags_keep[pind],fontsize=22)\n",
    "\n",
    "fig.legend(['truth','8-step DMD \\n prediction'],frameon=False,loc=(0.61,0.1),fontsize=22);\n",
    "fig.text(0.5, -0.02, 'Time after malathion induction (minutes)', ha='center',fontsize=27);\n",
    "fig.text(-0.03, 0.5, 'Standardized Fold Change', va='center', rotation='vertical',fontsize=29);\n",
    "fig.text(0.72,0.256,r'$R^2=0.92$' '\\n (all genes)',ha='center',fontsize=22)\n",
    "plt.tight_layout()\n",
    "if savefig: \n",
    "    plt.savefig(figDir + 'fig2/error_timeseries.pdf',bbox_inches='tight',transparent=True,dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#### visualize the histogram of the per gene residuals\n",
    "\n",
    "savefig = False\n",
    "\n",
    "plt.rcParams['axes.spines.right'] = False\n",
    "plt.rcParams['axes.spines.top'] = False \n",
    "\n",
    "mse_per_gene = (np.linalg.norm(np.linalg.norm(Xpred - data_fc_norm,axis=2),axis=1) \n",
    "                / (Xpred.shape[1]*Xpred.shape[2]))\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6,5))\n",
    "ax.xaxis.set_tick_params(which='both', size=6, width=1.5, direction='out',labelsize=25)\n",
    "ax.yaxis.set_tick_params(which='both', size=6, width=1.5, direction='out',labelsize=25)\n",
    "sns.histplot(mse_per_gene)\n",
    "plt.xlabel('Mean squared error',fontsize=27)\n",
    "plt.ylabel('Genes',fontsize=29)\n",
    "if savefig: \n",
    "    plt.savefig(figDir + 'fig2/mse_histogram_dmd.pdf',bbox_inches='tight',transparent=True,dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Subfigures for Figure 3 in manuscript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Xo1, Xo1_proj = finite_horizon_obs_gramian(A,V[:,0:1].T,10)\n",
    "print('Xo1 volume:', np.linalg.det(Xo1_proj))\n",
    "Xo2, Xo2_proj = finite_horizon_obs_gramian(A,V[:,1:2].T,10)\n",
    "print('Xo2 volume:', np.linalg.det(Xo2_proj))\n",
    "Xo3, Xo3_proj = finite_horizon_obs_gramian(A,V[:,2:3].T,10)\n",
    "print('Xo3 volume:', np.linalg.det(Xo3_proj))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Xo, Xo_proj = finite_horizon_obs_gramian(A,V[:,0:3].T,10)\n",
    "print('Xo volume:', np.linalg.det(Xo_proj))\n",
    "# though volume is extremely large when using multiple outputs, can't algorithmically use the multiple outputs\n",
    "# to identify which genes to select as sensors since the columns and rows of the sampling matrix have \n",
    "# do not have just a single '1' for yes do select and '0' everywhere else to say do not select. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# plot gramian ellipsoids\n",
    "savefig = False\n",
    "\n",
    "X1,Y1,Z1 = get_ellipsoid(Xo1_proj)\n",
    "X2,Y2,Z2 = get_ellipsoid(Xo2_proj)\n",
    "X3,Y3,Z3 = get_ellipsoid(Xo3_proj)\n",
    "\n",
    "fig, ax = plt.subplots(1,1,figsize=(6,6),subplot_kw=dict(projection='3d'))\n",
    "\n",
    "ax.plot_wireframe(np.zeros_like(X1),np.zeros_like(Y1),np.zeros_like(Z1),color='tab:blue',label=r'$\\mathcal{X}_o(\\mathbf{w}_1)$')\n",
    "ax.plot_wireframe(np.zeros_like(X1),np.zeros_like(Y1),np.zeros_like(Z1),color='tab:orange',label=r'$\\mathcal{X}_o(\\mathbf{w}_2)$')\n",
    "ax.plot_wireframe(np.zeros_like(X1),np.zeros_like(Y1),np.zeros_like(Z1),color='tab:green',label=r'$\\mathcal{X}_o(\\mathbf{w}_3)$')\n",
    "ax.plot_wireframe(X1,Y1,Z1, color='tab:blue', alpha=0.1)\n",
    "ax.plot_wireframe(X2,Y2,Z2, color='tab:orange', alpha=0.15)\n",
    "ax.plot_wireframe(X3,Y3,Z3, color='tab:green', alpha=0.2)\n",
    "ax.set_xlabel('Principal axis 1',labelpad=10)\n",
    "ax.set_ylabel('Principal axis 2',labelpad=10)\n",
    "ax.set_zlabel('Principal axis 3',labelpad=10)\n",
    "# get rid of colored axes planes\n",
    "ax.xaxis.pane.fill = False\n",
    "ax.yaxis.pane.fill = False\n",
    "ax.zaxis.pane.fill = False\n",
    "ax.xaxis.pane.set_edgecolor('w')\n",
    "ax.yaxis.pane.set_edgecolor('w')\n",
    "ax.zaxis.pane.set_edgecolor('w')\n",
    "ax.set_ylim(-3,3)\n",
    "ax.set_xlim(-7,7)\n",
    "# ax.set_zlim(-2,2)\n",
    "plt.legend(frameon=False,ncol=2,bbox_to_anchor=(1.1,1.00),fontsize=23)\n",
    "if savefig: \n",
    "    plt.savefig(figDir + 'fig3/gramian_ellipsoids.pdf',dpi=300,bbox_inches='tight',transparent=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sampling weights unscaled\n",
    "savefig = False\n",
    "\n",
    "plt.rcParams['axes.spines.right'] = False\n",
    "plt.rcParams['axes.spines.top'] = False \n",
    "\n",
    "fig,ax = plt.subplots(1,1,figsize=(5,4))\n",
    "plt.plot(np.sort(np.abs(W[0])))\n",
    "\n",
    "ax.xaxis.set_tick_params(which='both', size=6, width=1.5, direction='out')\n",
    "ax.yaxis.set_tick_params(which='both', size=6, width=1.5, direction='out')\n",
    "ax.set_xlabel('Gene ',fontsize=27);\n",
    "ax.set_ylabel('Sampling weights',fontsize=29);\n",
    "\n",
    "if savefig:\n",
    "    plt.savefig(figDir + 'supplementalFigs/Sampling_weights.pdf',dpi=300,bbox_inches='tight',transparent=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# visualize the weights in gene sampling vector and form three groups\n",
    "savefig = False\n",
    "\n",
    "plt.rcParams['axes.spines.right'] = False\n",
    "plt.rcParams['axes.spines.top'] = False \n",
    "\n",
    "# W_scaled = (np.abs(W[0]) - np.abs(W[0]).min())/(np.abs(W[0]).max() - np.abs(W[0]).min())\n",
    "\n",
    "# # get the indices of the genes with the ngenes/3 largest sampling weights\n",
    "nInds = int(len(data_fc_norm)/3)\n",
    "# high_inds = list(np.abs(W[0]).argsort()[-nInds:][::-1])\n",
    "# low_inds = list(np.abs(W[0]).argsort()[:nInds])\n",
    "# mid_inds = list(set(list(range(len(W[0])))) - set(high_inds).union(set(low_inds)))\n",
    "\n",
    "fig,ax = plt.subplots(1,1,figsize=(5,4))\n",
    "# ax.plot(range(nInds),np.abs(W[0][low_inds]),mew=0.3,ms=10,c='tab:blue',alpha=1)\n",
    "# ax.plot(range(nInds,2*nInds),np.sort(np.abs(W[0][mid_inds])),mew=0.3,ms=10,c='tab:orange',alpha=1)\n",
    "# ax.plot(range(2*nInds,3*nInds),np.abs(W[0][high_inds])[::-1],mew=0.3,ms=10,c='tab:green',alpha=1)\n",
    "ax.plot(range(nInds),np.sort(np.abs(Wn))[0:nInds])\n",
    "ax.plot(range(nInds,2*nInds),np.sort(np.abs(Wn))[nInds:2*nInds])\n",
    "ax.plot(range(2*nInds,3*nInds),np.sort(np.abs(Wn))[2*nInds:3*nInds])\n",
    "# ax.plot([0,len(Ws)],[np.abs(Wn).min(),np.abs(Wn).max()],'--',c='k',lw=1.5,zorder=0)\n",
    "\n",
    "ax.xaxis.set_tick_params(which='both', size=6, width=1.5, direction='out')\n",
    "ax.yaxis.set_tick_params(which='both', size=6, width=1.5, direction='out')\n",
    "ax.set_xlabel('Gene ',fontsize=27);\n",
    "# ax.set_ylabel('Sampling weights',fontsize=29);\n",
    "ax.set_ylabel('Sampling weights \\n (scaled)',fontsize=29);\n",
    "\n",
    "if savefig:\n",
    "    plt.savefig(figDir + 'fig3/Sampling_weights.pdf',dpi=300,bbox_inches='tight',transparent=True)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# get the indices of the genes with the ngenes/3 largest sampling weights\n",
    "nInds = int(len(data_fc_norm)/3)\n",
    "high_inds = list(np.abs(Wn).argsort()[-nInds:][::-1])\n",
    "low_inds = list(np.abs(Wn).argsort()[:nInds])\n",
    "mid_inds = list(set(list(range(len(Wn)))) - set(high_inds).union(set(low_inds)))\n",
    "\n",
    "def mc_reconstruction(p=50,nT=1,N=100):\n",
    "    '''\n",
    "    p: number of genes to sample per run\n",
    "    nT: number of timepoints to sample each gene for\n",
    "    N: number of runs\n",
    "    '''\n",
    "    r2_high = []\n",
    "    for ii in range(N):\n",
    "        inds = choices(high_inds,k=p)\n",
    "        set_to_zero = list(set(list(range(len(Wn)))) - set(inds) )\n",
    "        C = deepcopy(Wn)\n",
    "        C = np.expand_dims(C,axis=0)\n",
    "        C[:,set_to_zero] = 0.0\n",
    "        r2_high.append(reconstruct_x0(data_fc_norm,nT,A,C))\n",
    "\n",
    "    r2_mid = []\n",
    "    for ii in range(N):\n",
    "        inds = choices(mid_inds,k=p)\n",
    "        set_to_zero = list(set(list(range(len(Wn)))) - set(inds) )\n",
    "        C = deepcopy(Wn)\n",
    "        C = np.expand_dims(C,axis=0)\n",
    "        C[:,set_to_zero] = 0.0\n",
    "        r2_mid.append(reconstruct_x0(data_fc_norm,nT,A,C))\n",
    "\n",
    "    r2_low = []\n",
    "    for ii in range(N):\n",
    "        inds = choices(low_inds,k=p)\n",
    "        set_to_zero = list(set(list(range(len(Wn)))) - set(inds) )\n",
    "        C = deepcopy(Wn)\n",
    "        C = np.expand_dims(C,axis=0)        \n",
    "        C[:,set_to_zero] = 0.0\n",
    "        r2_low.append(reconstruct_x0(data_fc_norm,nT,A,C))\n",
    "        \n",
    "    return r2_low, r2_mid, r2_high\n",
    "\n",
    "# takes several minutes to run\n",
    "doRun = False\n",
    "doDump = False\n",
    "if doRun:\n",
    "    r2_low_nTlow, r2_mid_nTlow, r2_high_nTlow = mc_reconstruction(p=50,nT=2,N=100) # p=nInds//2\n",
    "    r2_low, r2_mid, r2_high = mc_reconstruction(p=50,nT=8,N=100) # p=nInds//2 (50)     \n",
    "    mc_recon_list = [r2_low_nTlow, r2_mid_nTlow, r2_high_nTlow, r2_low, r2_mid, r2_high]\n",
    "\n",
    "if doDump: \n",
    "    pickle.dump(mc_recon_list,open('data/mc_reconstruction.pkl','wb'))    \n",
    "    \n",
    "if not doRun:    \n",
    "    mc_recon_list = pickle.load(open('data/mc_reconstruction.pkl','rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mc_reconstruction_single(p=50,nT=1,N=100):\n",
    "    '''\n",
    "    p: number of genes to sample per run\n",
    "    nT: number of timepoints to sample each gene for\n",
    "    N: number of runs\n",
    "    '''\n",
    "\n",
    "    r2_high = []\n",
    "    for ii in range(N):\n",
    "        inds = choices(high_inds,k=p)\n",
    "        set_to_zero = list(set(list(range(len(Wn)))) - set(inds) )\n",
    "        C = deepcopy(Wn)\n",
    "        C = np.expand_dims(C,axis=0)\n",
    "        C[:,set_to_zero] = 0.0\n",
    "        r2_high.append(reconstruct_x0(data_fc_norm,nT,A,C))\n",
    "    \n",
    "    return r2_high\n",
    "\n",
    "\n",
    "doRun = False\n",
    "doDump = False\n",
    "if doRun:\n",
    "    times = [1,2,3,4,5,6,7,8,9,10]\n",
    "    r2_times_high = []\n",
    "    for this_t in times: \n",
    "        r2_high = mc_reconstruction_single(p=50,nT=this_t,N=50)\n",
    "        r2_times_high.append( r2_high )\n",
    "\n",
    "if doDump:\n",
    "    pickle.dump(r2_times_high, open('data/mc_reconstruction_highs.pkl','wb'))\n",
    "    \n",
    "if not doRun:\n",
    "    r2_times_high = pickle.load(open('data/mc_reconstruction_highs.pkl','rb'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "savefig = False\n",
    "\n",
    "times = [1,2,3,4,5,6,7,8,9,10]\n",
    "\n",
    "\n",
    "fig,ax=plt.subplots(2,1,figsize=(4.5,8.5),sharex=False,sharey=False) # gridspec_kw={'height_ratios': [1, 1.5]}\n",
    "sns.stripplot(data=pd.DataFrame(mc_recon_list[0:3]).T,s=6,linewidth=1,ax=ax[0])\n",
    "sns.stripplot(data=pd.DataFrame(r2_times_high).T,color='tab:green',s=6,linewidth=0.5,ax=ax[1])\n",
    "\n",
    "fig.text(-0.18, 0.5, 'Reconstruction accuracy', va='center', rotation='vertical',fontsize=29)\n",
    "ax[0].set_title('T=2',fontsize=22)\n",
    "for thisax in ax.reshape(-1):\n",
    "    thisax.xaxis.set_tick_params(which='both', size=6, width=1.5, direction='out')\n",
    "    thisax.yaxis.set_tick_params(which='both', size=6, width=1.5, direction='out')\n",
    "ax[0].set_xticklabels(['low','mid','high'],fontsize=24)\n",
    "ax[1].set_xticklabels(times,fontsize=24)\n",
    "ax[1].set_xlabel('T',fontsize=27)\n",
    "if savefig:\n",
    "    plt.savefig(figDir + 'fig3/monte_carlo_reconstruction.pdf',dpi=300,bbox_inches='tight',transparent=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# histogram of the weights in C\n",
    "savefig = False\n",
    "\n",
    "plt.rcParams['axes.spines.right'] = False\n",
    "plt.rcParams['axes.spines.top'] = False \n",
    "\n",
    "fig,ax = plt.subplots(1,1,figsize=(5,3))\n",
    "# ax.hist((np.abs(Wsorted[0]) - np.abs(Wsorted[0]).min())/(np.abs(Wsorted[0]).max() - np.abs(Wsorted[0]).min()),\n",
    "#         rwidth=0.825,align='right',linewidth=1.5,edgecolor='k',alpha=0.9)\n",
    "# sns.histplot((np.abs(Wsorted[0]) - np.abs(Wsorted[0]).min())/(np.abs(Wsorted[0]).max() - np.abs(Wsorted[0]).min()),kde=True,ax=ax)\n",
    "sns.histplot(np.abs(Wn),kde=True)\n",
    "# plt.plot([0.05,0.05],[0,100],c='r')\n",
    "\n",
    "ax.xaxis.set_tick_params(which='both', size=6, width=1.5, direction='out')\n",
    "ax.yaxis.set_tick_params(which='both', size=6, width=1.5, direction='out')\n",
    "ax.set_xlabel('Sampling weights',fontsize=27);\n",
    "ax.set_ylabel('Genes',fontsize=29);\n",
    "if savefig:\n",
    "    plt.savefig(figDir + 'fig3/sampling_weights_hist.pdf',dpi=300,bbox_inches='tight',transparent=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# state estimation error\n",
    "savefig = False\n",
    "\n",
    "plt.rcParams['axes.spines.right'] = False\n",
    "plt.rcParams['axes.spines.top'] = False \n",
    "plt.rc('lines',linewidth=4);\n",
    "plt.rc('lines',markersize=16);\n",
    "\n",
    "cmap = matplotlib.cm.get_cmap('YlGn_r')\n",
    "grad_colors = [list(cmap(x)[0:3]) for x in np.linspace(0.2,0.9,10)]\n",
    "fig,ax = plt.subplots(1,1,figsize=(9,8));\n",
    "mec = 'k'\n",
    "mew = 0.75\n",
    "alpha = 1\n",
    "xrange = np.linspace(samplingFreq,620,len(rho_list[0]))\n",
    "my_labels = [r'T=$10$',r'T=$9$',r'T=$8$',r'T=$7$',r'T=$6$',r'T=$5$',r'T=$4$',r'T=$3$',r'T=$2$',r'T=$1$']\n",
    "my_markers = ['o','s','D','^','v','<','>','p','d','X']\n",
    "for ii in range(len(rho_list)):\n",
    "    ax.plot(xrange,rho_list[ii],my_markers[ii]+'-',mfc=grad_colors[ii]+[alpha],c=grad_colors[ii]+[alpha],\n",
    "            mec=mec,mew=mew,ms=11,lw=2,label=my_labels[ii],zorder=10-ii)\n",
    "# ax.set_xticks([1,20,40,60,80])\n",
    "ax.set_xlabel('Genes (sampled in order of rank)',fontsize=27);\n",
    "ax.set_ylabel('Reconstruction accuracy',fontsize=29);\n",
    "# ax.set_ylabel(r'$\\mathregular{\\rho(\\mathbf{z}_0,\\mathcal{O}_T^{\\dagger} \\mathbf{y})}$',fontsize=26);\n",
    "ax.legend(frameon=False,ncol=2,fontsize=20)\n",
    "ax.xaxis.set_tick_params(which='both', size=6, width=1.5, direction='out')\n",
    "ax.yaxis.set_tick_params(which='both', size=6, width=1.5, direction='out')\n",
    "if savefig: \n",
    "    plt.savefig(figDir + 'fig3/state_estimation_error.pdf',dpi=300,bbox_inches='tight',transparent=True)    \n",
    "plt.show();"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # state estimation error compared with random sampling\n",
    "# savefig = False\n",
    "\n",
    "# plt.rcParams['axes.spines.right'] = False\n",
    "# plt.rcParams['axes.spines.top'] = False \n",
    "# plt.rc('lines',linewidth=3);\n",
    "\n",
    "# fig,axs = plt.subplots(2,5,figsize=(19,7),sharex=True,sharey=True);\n",
    "# mec = 'k'\n",
    "# mew = 1\n",
    "# alpha = 0.8\n",
    "# xrange = np.linspace(samplingFreq,620,len(rho_list[0]))\n",
    "# my_labels = [r'T=$10$',r'T=$9$',r'T=$8$',r'T=$7$',r'T=$6$',r'T=$5$',r'T=$4$',r'T=$3$',r'T=$2$',r'T=$1$']\n",
    "# my_markers = ['o','s','D','^','v','<','>','p','d','X']\n",
    "# for ii,ax in enumerate(axs.reshape(-1)):\n",
    "#     ax.set_title(my_labels[ii])\n",
    "#     ax.plot(xrange,rho_list[ii],'k')\n",
    "#     ax.plot(xrange,rho_rand_list[ii],'--',c='k')\n",
    "#     for side in ax.spines.keys():\n",
    "#         ax.spines[side].set_linewidth(2.5)\n",
    "#     ax.xaxis.set_tick_params(which='both', size=5, width=2, direction='in')\n",
    "#     ax.yaxis.set_tick_params(which='both', size=5, width=2, direction='in')\n",
    "#     ax.grid(linestyle=':',linewidth=1.75,dashes=(2,4))\n",
    "# ax.legend(['ranked sampling','random sampling'],frameon=False,fontsize=12)\n",
    "# fig.text(0.5, 0.03, 'Number of genes', ha='center',fontsize=22);\n",
    "# fig.text(0.07, 0.5, 'Reconstruction error', va='center', rotation='vertical',fontsize=24);\n",
    "# if savefig: \n",
    "#     plt.savefig(figDir + 'supplementalFigs/rand_estimation_error.pdf',dpi=300,bbox_inches='tight',transparent=True)    \n",
    "# plt.show();"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mc_reconstruction_top_n(n=15,p=15,nT=2,N=100):\n",
    "    '''\n",
    "    n: top n% of genes only\n",
    "    p: number of genes to sample per run\n",
    "    nT: number of timepoints to sample each gene for\n",
    "    N: number of runs\n",
    "    '''\n",
    "    r2_list = []\n",
    "    corr_list = []\n",
    "    keep_inds = list(np.argsort(np.abs(Wn))[-int(len(Wn)*(n)/100):])\n",
    "    set_to_zero = list( set(list(range(len(Wn)))) - set(keep_inds) )\n",
    "    C = deepcopy(Wn)\n",
    "    C = np.expand_dims(C,axis=0)\n",
    "    C[:,set_to_zero] = 0.0\n",
    "    for ii in range(N):\n",
    "        Cn = deepcopy(C)\n",
    "        inds = choices(keep_inds,k=p)  \n",
    "        set_to_zero_n = list(set(keep_inds) - set(inds))\n",
    "        Cn[:,set_to_zero_n] = 0.0\n",
    "        r2_list.append(reconstruct_x0(data_fc_norm,nT,A,Cn))\n",
    "        corr_list.append( \n",
    "            np.linalg.norm(np.ones((p,p))-\n",
    "                           np.abs(np.corrcoef(data_fc_norm[inds].reshape(p,-1,order='F'))),ord='fro') )\n",
    "    return r2_list, corr_list\n",
    "\n",
    "doRun = False\n",
    "doDump = False\n",
    "if doRun:\n",
    "    r2_top_n, corr_top_n = mc_reconstruction_top_n(n=15,p=15,nT=nT,N=100)\n",
    "\n",
    "if doDump:\n",
    "    pickle.dump([r2_top_n,corr_top_n],open('data/top_n_list.pkl','wb'))\n",
    "\n",
    "if not doRun:\n",
    "    r2_top_n, corr_top_n = pickle.load(open('data/top_n_list.pkl','rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "fig,ax = plt.subplots(1,2,figsize=(12,6),gridspec_kw={'width_ratios':[0.4,0.6]})\n",
    "sns.boxplot(data=pd.DataFrame(r2_top_n),color='tab:gray',linewidth=2,ax=ax[0],\n",
    "            flierprops={\"marker\": \"o\",\"mfc\":'darkgray',\"mew\":2,\"ms\":8},\n",
    "            boxprops={'facecolor':'none', 'edgecolor':'k'})\n",
    "ax[0].set_xticklabels([''],fontsize=24)\n",
    "ax[0].set_xlabel('Random set of 15 \\n from top 15% of \\n observability ranking',fontsize=23)\n",
    "ax[0].set_ylabel('Reconstruction accuracy',fontsize=27.5)\n",
    "sns.scatterplot(x=r2_top_n,y=corr_top_n,ax=ax[1],color='darkgray',edgecolor='k',s=100)\n",
    "# chosen gene set \n",
    "plt.plot(0.5080958828078517,7.0,'o',c='red',mec='k',ms=10)\n",
    "ax[1].text(0.53,6.55,'malathion reporter \\n library',fontsize=22)\n",
    "ax[1].set_xlabel('Reconstruction accuracy')\n",
    "ax[1].set_ylabel('Correlation metric\\n' + \n",
    "                 r'$|| \\mathbf{1}_{k\\times k} - \\mathbf{R}^{\\mathrm{abs}}||_F $',fontsize=27.5)\n",
    "ax[0].set_ylim(0.1,0.7)\n",
    "ax[1].set_ylim(2.5,8)\n",
    "ax[1].set_xlim(0.18,0.7)\n",
    "plt.tight_layout()\n",
    "# plt.savefig(figDir + 'supplementalFigs/mc_analysis_top.pdf',bbox_inches='tight',transparent=True,dpi=300);\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mc_reconstruction_random(n=300,p=15,nT=2,N=100):\n",
    "    '''\n",
    "    n: number of genes to subsample from\n",
    "    p: number of genes to sample per run\n",
    "    nT: number of timepoints to sample each gene for\n",
    "    N: number of runs\n",
    "    '''\n",
    "    r2_list = []\n",
    "    corr_list = []\n",
    "    rank_list = [] # average rank of gene set\n",
    "    inds1 = choices(list(range(len(Wn))),k=n)\n",
    "    for ii in range(N):\n",
    "        inds2 = choices(inds1,k=p)\n",
    "        set_to_zero = list(set(list(range(len(Wn)))) - set(inds2))\n",
    "        C = deepcopy(Wn)\n",
    "        C = np.expand_dims(C,axis=0)\n",
    "        C[:,set_to_zero] = 0.0\n",
    "        r2_list.append(reconstruct_x0(data_fc_norm,nT,A,C))\n",
    "        corr_list.append( \n",
    "            np.linalg.norm(np.ones((p,p))-\n",
    "                           np.abs(np.corrcoef(data_fc_norm[inds2].reshape(p,-1,order='F'))),ord='fro') )\n",
    "        rank_list.append( np.mean([list(np.argsort(np.abs(Wn))).index(x) for x in inds2]))\n",
    "    return r2_list, corr_list, rank_list\n",
    "\n",
    "doRun = False\n",
    "doDump = False\n",
    "if doRun:\n",
    "    r2_random, corr_random, rank_random = mc_reconstruction_random(n=len(Wn)//2,p=11,nT=6,N=100)\n",
    "    \n",
    "if doDump: \n",
    "    pickle.dump([r2_random,corr_random,rank_random],open('data/random_list.pkl','wb'))\n",
    "\n",
    "if not doRun:\n",
    "    r2_random, corr_random, rank_random = pickle.load(open('data/random_list.pkl','rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax = plt.subplots(1,2,figsize=(12,6),gridspec_kw={'width_ratios':[0.4,0.6]})\n",
    "sns.boxplot(data=pd.DataFrame(r2_random),color='tab:gray',linewidth=2,ax=ax[0],\n",
    "            flierprops={\"marker\": \"o\",\"mfc\":'darkgray',\"mew\":2,\"ms\":8},\n",
    "            boxprops={'facecolor':'none', 'edgecolor':'k'})\n",
    "ax[0].set_xticklabels([''],fontsize=24)\n",
    "ax[0].set_xlabel('Random set of 15 \\n from random 50% of \\n observability ranking',fontsize=24)\n",
    "ax[0].set_ylabel('Reconstruction accuracy',fontsize=27.5)\n",
    "sns.scatterplot(x=r2_random,y=corr_random,ax=ax[1],edgecolor='k',color='darkgray',s=100)\n",
    "# chosen gene set \n",
    "plt.plot(0.5080958828078517,7.0,'o',c='red',mec='k',ms=10)\n",
    "# ax[1].text(0.53,6.55,'malathion reporter \\n library',fontsize=22)\n",
    "ax[1].set_xlabel('Reconstruction accuracy'); \n",
    "ax[1].set_ylabel('Correlation metric\\n' + \n",
    "                 r'$|| \\mathbf{1}_{k\\times k} - \\mathbf{R}^{\\mathrm{abs}}||_F $',fontsize=27.5)\n",
    "ax[0].set_ylim(0.1,0.7)\n",
    "ax[1].set_ylim(2.5,8)\n",
    "ax[1].set_xlim(0.18,0.7)\n",
    "plt.tight_layout()\n",
    "# plt.savefig(figDir + 'supplementalFigs/mc_analysis_random.pdf',bbox_inches='tight',transparent=True,dpi=300);\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reconstruct x0 with only the chosen sensor genes\n",
    "set_to_zero = list(set(list(range(len(Wn)))) - set(list(df_up_new.tag_ind)))\n",
    "C = deepcopy(Wn)\n",
    "C[set_to_zero] = 0.0\n",
    "C = C.reshape(-1,len(C))\n",
    "r2 = reconstruct_x0(data_fc_norm,nT,A,C)\n",
    "print(r2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reconstruction x0 using top 15 genes in Wn\n",
    "set_to_zero = list(set(list(range(len(Wn)))) - set(list(np.argsort(np.abs(Wn))[-15:])))\n",
    "C = deepcopy(Wn)\n",
    "C[set_to_zero] = 0.0\n",
    "C = C.reshape(-1,len(C))\n",
    "r2 = reconstruct_x0(data_fc_norm,nT,A,C)\n",
    "print(r2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# correlations among chosen sensor genes\n",
    "\n",
    "cg = sns.clustermap(np.corrcoef(data_fc_norm[list(df_up_new.tag_ind)].reshape(15,-1,order='F')),\n",
    "           vmax=1,vmin=-1,cmap='bwr',cbar_kws={'label':'Pearson correlation'},figsize=(8,8),)\n",
    "reordered_inds = cg.dendrogram_col.reordered_ind\n",
    "cg.ax_row_dendrogram.set_visible(False) #suppress row dendrogram\n",
    "cg.ax_col_dendrogram.set_visible(False) #suppress column dendrogram\n",
    "cg.ax_heatmap.set_xticks(range(len(df_up_new)))\n",
    "cg.ax_heatmap.set_xticklabels([''.join(list(df_up_new.gene)[x].split('\\n ')) for x in reordered_inds],\n",
    "                             rotation=90,fontsize=28)\n",
    "cg.ax_heatmap.set_yticks(range(len(df_up_new)))\n",
    "cg.ax_heatmap.set_yticklabels([''.join(list(df_up_new.gene)[x].split('\\n ')) for x in reordered_inds],\n",
    "                             rotation=0,fontsize=27)\n",
    "# cg.fig.savefig(figDir + 'supplementalFigs/library_correlations.pdf',dpi=300,transparent=True,bbox_inches='tight')\n",
    "\n",
    "print( np.linalg.norm( np.ones((15,15)) - np.abs(np.corrcoef(data_fc_norm[list(df_up_new.tag_ind)].reshape(15,-1,order='F'))), ord='fro' ) )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Top 15 genes ranked by Wn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# protein name, molecular function, biological process\n",
    "\n",
    "df_top_genes = pd.DataFrame({'locus_tag': [locus_tags_keep[x] for x in sorted_inds_n[::-1][:15]],\n",
    " 'gene': [genes_keep[x] for x in sorted_inds_n[::-1][:15]],\n",
    " 'protein_name':['50S ribosomal protein L34',\n",
    "                'Putative outer membrane protein',\n",
    "                'ATP-dependent Clp protease proteolytic subunit',\n",
    "                'Translational regulator CsrA',\n",
    "                'ATP synthase subunit a',\n",
    "                '30S ribosomal protein S11',\n",
    "                '50S ribosomal protein L6',\n",
    "                '50S ribosomal protein L10',\n",
    "                'ATP synthase subunit b',\n",
    "                'Coenzyme PQQ synthesis protein A',\n",
    "                'Ribose-phosphate pyrophosphokinase',\n",
    "                'Probable Fe(2+)-trafficking protein',\n",
    "                '30S ribosomal protein S7',\n",
    "                'Arginine deiminase',\n",
    "                '50S ribosomal protein L16'],\n",
    "'molecular function':['structural constituent of ribosome',\n",
    "                      'N/A',\n",
    "                      'ATP-dependent peptidase activity; serine-type endopeptidase activity',\n",
    "                      \"mRNA 5'-UTR binding\",\"proton-transporting ATP synthase activity, rotational mechanism\",\n",
    "                      'structural constituent of ribosome',\n",
    "                      'structural constituent of ribosome',\n",
    "                      'structural constituent of ribosome',\n",
    "                      'hydrolase activity; proton-transporting ATP synthase activity, rotational mechanism',\n",
    "                      'N/A',\n",
    "                      'ATP binding; kinase activity; magnesium ion binding; ribose phosphate diphosphokinase activiity',\n",
    "                      'N/A',\n",
    "                      'rRNA binding; structural constituent of ribosome; tRNA binding',\n",
    "                      'arginine deiminase activity',\n",
    "                      'rRNA binding; structural constituent of ribosome; tRNA binding'],\n",
    "'biological process':['translation',\n",
    "                     'N/A',\n",
    "                     'proteolysis',\n",
    "                     'regulation of translational initation and carbohydrate metabolic process',\n",
    "                     'N/A',\n",
    "                     'translation',\n",
    "                     'translation',\n",
    "                     'translation',\n",
    "                     'N/A',\n",
    "                     'pyrroloquinoline quinone biosynthetic process',\n",
    "                     '5-phosphoribose 1-diphosphate biosynthetic process; nucleotide biosynthetic process; phosphorylation; ribonucleoside monophosphate biosynthetic process',\n",
    "                     'N/A',\n",
    "                     'translation',\n",
    "                     'arginine catabolic process to  ornithine',\n",
    "                     'translation'] })\n",
    "\n",
    "for ii in range(len(df_top_genes)):\n",
    "    if df_top_genes.gene[ii] == 'N/A':\n",
    "        df_top_genes.gene[ii] = df_top_genes.protein_name[ii]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# correlations among top 15 by Wn\n",
    "cg = sns.clustermap(np.corrcoef(data_fc_norm[list(np.argsort(np.abs(Wn))[-15:])].reshape(15,-1,order='F')),\n",
    "           vmax=1,vmin=-1,cmap='bwr',cbar_kws={'label':'Pearson correlation'},figsize=(8,8),)\n",
    "reordered_inds = cg.dendrogram_col.reordered_ind\n",
    "cg.ax_row_dendrogram.set_visible(False) #suppress row dendrogram\n",
    "cg.ax_col_dendrogram.set_visible(False) #suppress column dendrogram\n",
    "cg.ax_heatmap.set_xticks(range(len(df_up_new)))\n",
    "cg.ax_heatmap.set_xticklabels([list(df_top_genes.gene)[x] for x in reordered_inds],rotation=90,\n",
    "                             fontsize=28)\n",
    "cg.ax_heatmap.set_yticks(range(len(df_up_new)))\n",
    "cg.ax_heatmap.set_yticklabels([list(df_top_genes.gene)[x] for x in reordered_inds],rotation=0,\n",
    "                             fontsize=27)\n",
    "# cg.fig.savefig(figDir + 'supplementalFigs/top_correlations.pdf',dpi=300,transparent=True,bbox_inches='tight')\n",
    "\n",
    "print( np.linalg.norm( np.ones((15,15)) - np.abs(np.corrcoef(data_fc_norm[list(np.argsort(np.abs(Wn))[-15:])].reshape(15,-1,order='F'))), ord='fro' ) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_top_genes.to_csv(figDir+'supplementalFigs/top_genes.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Correlation strategy for gene selection -- accounts for observability by starting at the highest ranked gene and proceeding from there"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# sort genes in data matrix by descending rank and reshape data matrix to 2D (genes x condition)\n",
    "# corr_cutoff = 0.1\n",
    "corr_genes_r2s = []\n",
    "corr_gene_sets = [] \n",
    "corr_gene_set_norms = []\n",
    "# for corr_cutoff in np.linspace(0.1,0.9,50):\n",
    "# for corr_cutoff in [0.73157895]:\n",
    "for corr_cutoff in [0.5]: \n",
    "    corr_data = data_fc_norm[sorted_inds_n[::-1]].reshape(len(data_fc_norm),-1,order='F')\n",
    "    corr = np.abs(np.corrcoef(corr_data))\n",
    "\n",
    "    corr[corr >= corr_cutoff] = 1.0\n",
    "\n",
    "    corr_gene_list = [sorted_inds_n[::-1][0]]\n",
    "    this_row = 0\n",
    "    while len(corr_gene_list) < 15: \n",
    "        # get the value, index of the first col of this_row that is less than 1.0\n",
    "        next_ind = [n for n, this_corr in enumerate(corr[this_row,this_row:]) if this_corr < 1.0][0] \n",
    "        this_row = next_ind + this_row\n",
    "        corr_gene_list.append(sorted_inds_n[::-1][this_row]) # these are the indices in data_fc_norm corresponding to the chosen genes\n",
    "    corr_gene_sets.append(corr_gene_list)\n",
    "\n",
    "    # reconstruction accuracy with these selected genes\n",
    "    set_to_zero = list(set(list(range(len(Wn)))) - set(corr_gene_list))\n",
    "    C = deepcopy(Wn)\n",
    "    C[set_to_zero] = 0.0\n",
    "    C = C.reshape(-1,len(C))\n",
    "    r2 = reconstruct_x0(data_fc_norm,nT,A,C)\n",
    "    corr_genes_r2s.append(r2)\n",
    "    \n",
    "    # norm of fold change response for the genes in the gene set\n",
    "    corr_gene_set_norms.append(\n",
    "        np.linalg.norm(data_fc.reshape(len(data_fc),-1,order='F')[corr_gene_list]) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "corr_gene_sets[0], corr_genes_r2s[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print( np.linalg.norm( np.ones((15,15)) - np.abs(np.corrcoef(data_fc_norm[corr_gene_sets[0]].reshape(15,-1,order='F'))), ord='fro' ) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reconstruct x0 with only the chosen sensor genes\n",
    "set_to_zero = list(set(list(range(len(Wn)))) - set(corr_gene_sets[0]))\n",
    "C = deepcopy(Wn)\n",
    "C[set_to_zero] = 0.0\n",
    "C = C.reshape(-1,len(C))\n",
    "r2 = reconstruct_x0(data_fc_norm,nT,A,C)\n",
    "print(r2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "corr_gene_names = [genes_keep[x] if genes_keep[x] != 'N/A' else locus_tags_keep[x] for x in corr_gene_sets[0]]\n",
    "corr_ranks = (np.array([sorted_inds_n.index(x) for x in corr_gene_sets[0]])+1)/624"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# plot corr_cutoff against reconstruction accuracy and norm of fold change response\n",
    "# plt.figure(figsize=(10, 8))\n",
    "# sns.scatterplot(x=np.linspace(0.1,0.9,50),y=corr_genes_r2s,c=corr_gene_set_norms)\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "plt.rcParams['axes.spines.right'] = False\n",
    "plt.rcParams['axes.spines.top'] = False\n",
    "plt.rcParams['axes.spines.bottom'] = True\n",
    "plt.rcParams['axes.spines.left'] = True\n",
    "\n",
    "xmin = 0\n",
    "xmax = 100\n",
    "fig, axs = plt.subplots(3,5,figsize=(18,10),sharex=True); \n",
    "for ii,ax in enumerate(axs.reshape(-1)):\n",
    "    if ii < len(tags):\n",
    "        ax.set_title(corr_gene_names[ii],fontsize=20,x=0.75)\n",
    "        pBS = savgol(data_t_orig[keepers][corr_gene_sets[0][ii]] - data_c_orig[keepers][corr_gene_sets[0][ii]],5,2,axis=0)\n",
    "        mBS = np.mean(pBS,axis=1)\n",
    "        sBS = np.std(pBS,axis=1)\n",
    "        ax.errorbar(corr_tspan,mBS[2:],yerr=sBS[2:],fmt='o--',color='k',lw=2,\n",
    "                    ms=9,mfc='k',mec='k',mew=1.5,capsize=3,elinewidth=2,ecolor='k')\n",
    "        ax.set_xlabel(f'Rank: {100*corr_ranks[ii]:.2f}',fontsize=20,labelpad=2)\n",
    "        ax.set_xticks(range(10,90+40,40))\n",
    "        ax.hlines(0.0,xmin,xmax,color='r',linestyle='-.',lw=1.5)\n",
    "        ax.ticklabel_format(axis=\"y\", style=\"sci\",scilimits=(0,0),useMathText=True)\n",
    "        ax.yaxis.get_offset_text().set_fontsize(22)\n",
    "        ax.xaxis.set_tick_params(which='both', size=6, width=1.5, direction='out',labelsize=22)\n",
    "        ax.yaxis.set_tick_params(which='both', size=6, width=1.5, direction='out',labelsize=22)\n",
    "        ymin,ymax = ax.get_ylim()\n",
    "\n",
    "fig.text(0.5, 0.01, 'Time after malathion induction (minutes)', ha='center',fontsize=27);\n",
    "fig.text(0.065, 0.5, 'Background subtracted TPM', va='center', rotation='vertical',fontsize=29);\n",
    "plt.subplots_adjust(wspace=0.35, hspace=0.3)\n",
    "# plt.savefig(figDir + 'supplementalFigs/corr_genes_rnaseq.pdf',bbox_inches='tight',transparent=True,dpi=300);\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# sensor gene TPMs\n",
    "savefig = False\n",
    "\n",
    "plt.rcParams['axes.spines.right'] = False\n",
    "plt.rcParams['axes.spines.top'] = False\n",
    "plt.rcParams['axes.spines.bottom'] = True\n",
    "plt.rcParams['axes.spines.left'] = True\n",
    "\n",
    "full_tspan = list(range(-10,100,10))\n",
    "\n",
    "cmap = matplotlib.cm.get_cmap('tab20')\n",
    "\n",
    "fig, axs = plt.subplots(3,5,figsize=(18,10),sharex=True); \n",
    "for ii,ax in enumerate(axs.reshape(-1)):\n",
    "    if ii < len(tags):\n",
    "        ax.set_title(df_up_new.gene[ii],fontsize=20,x=0.7)\n",
    "        pBS = savgol(data_t_orig[keepers][df_up_new.tag_ind[ii]] - data_c_orig[keepers][df_up_new.tag_ind[ii]],5,2,axis=0)\n",
    "        mBS = np.mean(pBS,axis=1)\n",
    "        sBS = np.std(pBS,axis=1)\n",
    "        ax.errorbar(full_tspan,mBS,yerr=sBS,fmt='o--',color='k',lw=2,\n",
    "                    ms=9,mfc=cmap(ii),mec='k',mew=1.5,capsize=3,elinewidth=2,ecolor='k')\n",
    "        ax.set_xlabel(f'Rank: {df_up_new.obs_stand_rank[ii]:.2f}',fontsize=20,labelpad=2)\n",
    "        ax.set_xticks(range(10,90+40,40))\n",
    "        ax.axvspan(-1,10,color='tab:gray',alpha=0.1)\n",
    "        ax.set_xlim(-0.5,)\n",
    "        xmin,xmax = ax.get_xlim()\n",
    "        ax.hlines(0.0,xmin,xmax,color='r',linestyle='-.',lw=1.5)\n",
    "        ax.ticklabel_format(axis=\"y\", style=\"sci\",scilimits=(0,0),useMathText=True)\n",
    "        ax.yaxis.get_offset_text().set_fontsize(22)\n",
    "        ax.xaxis.set_tick_params(which='both', size=6, width=1.5, direction='out',labelsize=22)\n",
    "        ax.yaxis.set_tick_params(which='both', size=6, width=1.5, direction='out',labelsize=22)\n",
    "        ymin,ymax = ax.get_ylim()\n",
    "        ax.vlines(10,ymin,ymax,color='darkgray',lw=2)\n",
    "        ax.set_ylim(ymin,ymax)\n",
    "\n",
    "fig.text(0.5, 0.01, 'Time after malathion induction (minutes)', ha='center',fontsize=27);\n",
    "fig.text(0.065, 0.5, 'Background subtracted TPM', va='center', rotation='vertical',fontsize=29);\n",
    "plt.subplots_adjust(wspace=0.52, hspace=0.7)\n",
    "\n",
    "if savefig: \n",
    "    plt.savefig(figDir + 'fig3/library_genes_rnaseq.pdf',bbox_inches='tight',transparent=True,dpi=300);\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Correlate the observability ranking and the gene's standard deviation\n",
    "\n",
    "savefig = False\n",
    "\n",
    "plt.rcParams['axes.spines.right'] = False\n",
    "plt.rcParams['axes.spines.top'] = False\n",
    "\n",
    "# gene_std = np.std(np.std(data_fc_norm,axis=2),axis=1)\n",
    "gene_std = np.mean(np.std(data_fc_norm,axis=2),axis=1)\n",
    "# W_scaled = (np.abs(W[0]) - np.abs(W[0]).min())/(np.abs(W[0]).max() - np.abs(W[0]).min()) \n",
    "W_scaled = np.abs(W[0])\n",
    "z, cov = np.polyfit(x=W_scaled, y=gene_std, deg=2, cov=True)\n",
    "# slope, intercept = z[0][0], z[0][1]\n",
    "# std_slope = np.sqrt(z[1][0,0])\n",
    "# std_intercept = np.sqrt(z[1][1,1])\n",
    "\n",
    "rhos, ps = sp.stats.spearmanr(np.sort(W_scaled), gene_std[np.argsort(W_scaled)])\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6,5))\n",
    "sns.scatterplot(x=np.sort(W_scaled), y=gene_std[np.argsort(W_scaled)], s=40)\n",
    "plt.plot(np.sort(W_scaled), z[0]*np.sort(W_scaled)**2+z[1]*np.sort(W_scaled)+z[2],color='tab:orange')\n",
    "# plt.fill_between(np.sort(W_scaled),\n",
    "#                  (z[0]-cov[0,0])*np.sort(W_scaled)**2+(z[1]-cov[1,1])*np.sort(W_scaled)+(z[2]-cov[2,2]),\n",
    "#                  (z[0]+cov[0,0])*np.sort(W_scaled)**2+(z[1]+cov[1,1])*np.sort(W_scaled)+(z[2]+cov[2,2]),\n",
    "#                  color='r',alpha=0.20)\n",
    "plt.xlabel('Sampling weights',fontsize=26); plt.ylabel('Standard deviation',fontsize=29)\n",
    "fig.text(1,0.65,\"Spearman's \\n\" + r\"$\\rho=$ \" + f\"{rhos:.2f}\",ha='center',fontsize=23)\n",
    "fig.text(1,0.55, r\"$p=$ \" + f\"{ps:.1e}\",ha='center',fontsize=23)\n",
    "\n",
    "    \n",
    "if savefig: \n",
    "    plt.savefig(figDir + 'fig3/correlate_obs_genestd.pdf',bbox_inches='tight',transparent=True,dpi=300);\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sp.stats.spearmanr(np.sort(W_scaled), gene_std[np.argsort(W_scaled)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Correlate the observability ranking and the gene's mean \n",
    "\n",
    "savefig = False\n",
    "\n",
    "plt.rcParams['axes.spines.right'] = False\n",
    "plt.rcParams['axes.spines.top'] = False\n",
    "\n",
    "gene_mean = np.mean(np.mean(data_fc_norm,axis=2),axis=1)\n",
    "W_scaled = (np.abs(W[0]) - np.abs(W[0]).min())/(np.abs(W[0]).max() - np.abs(W[0]).min()) \n",
    "z = np.polyfit(x=W_scaled, y=gene_mean, deg=1, cov=True)\n",
    "slope, intercept = z[0][0], z[0][1]\n",
    "std_slope = np.sqrt(z[1][0,0])\n",
    "std_intercept = np.sqrt(z[1][1,1])\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6,5))\n",
    "sns.scatterplot(x=np.sort(W_scaled), y=gene_mean[np.argsort(W_scaled)], s=40)\n",
    "plt.plot(np.sort(W_scaled), slope*np.sort(W_scaled)+intercept,color='red')\n",
    "plt.fill_between(np.sort(W_scaled),(slope-std_slope)*np.sort(W_scaled)+(intercept-std_intercept),\n",
    "                 (slope+std_slope)*np.sort(W_scaled)+(intercept+std_intercept),color='r',\n",
    "                alpha=0.20)\n",
    "plt.xlabel('Gene sampling weights'); plt.ylabel('Gene fold change (mean over time)')\n",
    "for ax_ in ['bottom','left']: \n",
    "    ax.spines[ax_].set_linewidth(2)\n",
    "    \n",
    "if savefig: \n",
    "    plt.savefig(figDir + 'fig3/correlate_obs_genemean.pdf',bbox_inches='tight',transparent=True,dpi=300);\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# differential expression comparisons\n",
    "\n",
    "deseq_df6 = pd.read_csv('data/deseq_t6.csv')\n",
    "deseq_df7 = pd.read_csv('data/deseq_t7.csv')\n",
    "deseq_df8 = pd.read_csv('data/deseq_t8.csv')\n",
    "deseq_df9 = pd.read_csv('data/deseq_t9.csv')\n",
    "deseq_df10 = pd.read_csv('data/deseq_t10.csv')\n",
    "deseq_df11 = pd.read_csv('data/deseq_t11.csv')\n",
    "\n",
    "deseq_files = ['deseq_t6','deseq_t7','deseq_t8','deseq_t9','deseq_t10','deseq_t11']\n",
    "li = []\n",
    "for file in deseq_files: \n",
    "    li.append(pd.read_csv('data/'+file + '.csv' ) )\n",
    "\n",
    "deseq_df = pd.concat(li, axis=0, ignore_index=True)\n",
    "# get union of tx_ids (due to some being repeated)\n",
    "deseq_ids = list(set( list(deseq_df.tx_id) ))\n",
    "\n",
    "# convert tx_ids to locus_tags\n",
    "de_genes, de_tags, _ = getRecords(cdsFasta_path,genbank_path,deseq_ids)\n",
    "\n",
    "# get the top n locus_tags from obs ranking, where n = len(deseq_ids)\n",
    "top_n_tags = [x for ii, x in enumerate(sorted_tags_n[::-1]) if ii < len(de_tags)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_in_common = set(top_n_tags).intersection(de_tags)\n",
    "print('genes in common between the dge analysis and the top n observability ranked genes')\n",
    "df_up_new[df_up_new.locus_tag.isin(genes_in_common)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(de_tags), len(top_n_tags)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(genes_in_common), len(set(de_tags)-set(top_n_tags)), len(set(top_n_tags)-set(de_tags))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib_venn import venn2, venn2_circles, venn2_unweighted\n",
    "\n",
    "savefig = False\n",
    "\n",
    "venn = venn2(subsets = (len(set(de_tags)-set(top_n_tags)), len(set(top_n_tags)-set(de_tags)), len(genes_in_common) ), \n",
    "      set_labels = (f'{len(de_tags)} differentially \\n expressed genes \\n' + r'(5 with $p_{\\mathrm{adj}} < 0.05$)',\n",
    "                    f'{len(de_tags)} genes \\n with largest \\n sampling weight'),\n",
    "            set_colors=('tab:orange', 'tab:blue'))\n",
    "\n",
    "\n",
    "venn.set_labels[0].set_x(0.1)\n",
    "venn.set_labels[1].set_x(0.2)\n",
    "for text in venn.set_labels:\n",
    "    text.set_fontsize(25)\n",
    "    \n",
    "for text in venn.subset_labels:\n",
    "    text.set_fontsize(25)\n",
    "\n",
    "venn2_circles(subsets = (len(set(de_tags)-set(top_n_tags)), \n",
    "                         len(set(top_n_tags)-set(de_tags)), len(genes_in_common) ),\n",
    "                         linewidth=1);\n",
    "\n",
    "if savefig: \n",
    "    plt.savefig(figDir + 'fig3/methods_overlap_venn.pdf',bbox_inches='tight',transparent=True,dpi=300);\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = pd.read_csv('data/tpm.csv')\n",
    "# getting transcriptome dynamics\n",
    "df_all = df_all[df_all.geneID.isin(deseq_ids)]\n",
    "txIDs_all = list(df_all.iloc[:,0])\n",
    "df_all = df_all.iloc[:,1:]\n",
    "# \n",
    "_, _, data_c_all, data_t_all, _, _, _, _, _, _ = process_df(df_all,sampleLabels,nreps,reps,ntimepts,txIDs_all)\n",
    "data_fc_all = (data_t_all + 20)/(data_c_all + 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# data_fc_fair = (data_t + 20 )/(data_c + 20)\n",
    "# data_fc_fair1 = data_fc_fair.reshape(len(data_fc_fair),data_fc_fair.shape[1]*data_fc_fair.shape[2],order='F')\n",
    "# scaler = StandardScaler()\n",
    "# scaler.fit(data_fc_fair1.T) \n",
    "# data_fc_fair1 = scaler.transform(data_fc_fair1.T).T\n",
    "# data_fc_fair = data_fc_fair1.reshape(len(data_fc_fair1),9,2,order='F')\n",
    "# A_fair,_,_,_,_,_,_,_,_,_ = \\\n",
    "#         dmd(data_fc_fair,rank_reduce=True,r=r,trim=False,trimThresh=2.5e-3) #trimThesh=2.5e-3\n",
    "# G_fair = gram_matrix(A_fair,data_fc_fair[:,0,:],nT=nT,reduced=False)\n",
    "# D_fair, V_fair = sp.linalg.eig(G_fair,left=False,right=True)\n",
    "# W_fair = V_fair[:,0:1].T\n",
    "# data_fc_fair_nonnorm = (data_t + 20 )/(data_c + 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "savefig = False\n",
    "\n",
    "plt.figure(figsize=(5.5,3.5))\n",
    "sns.histplot(np.linalg.norm(np.mean(data_fc[[locus_tags_keep.index(this_tag) for this_tag in top_n_tags]],\n",
    "                                    axis=2),axis=1),bins=15,kde=True,label='obs. max.',)\n",
    "# sns.histplot(np.linalg.norm(np.mean(data_fc_fair_nonnorm[np.argsort(np.abs(W_fair[0]))[-len(data_fc_all):]],axis=2),axis=1)\n",
    "#                           ,bins=15,kde=True,label='obs. max.',)\n",
    "sns.histplot(np.linalg.norm(np.mean(data_fc_all,axis=2),axis=1),bins=15,kde=True,color='tab:orange',label='DESeq2')\n",
    "# sns.kdeplot(np.linalg.norm(np.mean(data_fc[[locus_tags_keep.index(this_tag) for this_tag in top_n_tags]],axis=2)\n",
    "#                            ,axis=1),shade=True,label='obs. max.')\n",
    "# sns.kdeplot(np.linalg.norm(np.mean(data_fc_all,axis=2),axis=1),shade=True,color='tab:orange',label='DESeq2')\n",
    "plt.legend(frameon=False,fontsize=20)\n",
    "plt.xlabel('L2 norm of fold change responses',fontsize=25)\n",
    "plt.ylabel('Genes',fontsize=28)\n",
    "if savefig: \n",
    "    plt.savefig(figDir + 'fig3/compare_methods_hist.pdf',bbox_inches='tight',transparent=True,dpi=300);\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How about the genes that are differentially expressed after multiple-testing correction? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "de_corrected_genes, de_corrected_tags, _ = getRecords(cdsFasta_path,genbank_path,deseq_df[deseq_df.padj < 0.05].tx_id.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "de_corrected_id_inds = [txIDs.index(this_id) for this_id in deseq_df[deseq_df.padj < 0.05].tx_id.tolist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "inds_in_common_corrected = [locus_tags_keep.index(this_tag) for \n",
    "                             this_tag in de_corrected_tags if this_tag in locus_tags_keep]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "inds_in_common_corrected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tags_common = [locus_tags_keep[ii] for ii in inds_in_common_corrected]\n",
    "genes_common = [genes_keep[ii] for ii in inds_in_common_corrected]\n",
    "tags_common, genes_common"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rank of genes_common_corrected in obs. max. \n",
    "pr_deseq2 = [np.argsort(np.abs(Wn)).tolist().index(inds_in_common_corrected[ii])/(len(data_fc)) for ii in range(len(inds_in_common_corrected))]\n",
    "pr_deseq2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "common_dict = \\\n",
    "{this_tag:[locus_tags_keep.index(this_tag),de_corrected_id_inds[ii]] for ii, this_tag in enumerate(de_corrected_tags) if this_tag in locus_tags_keep}\n",
    "common_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(2,3,figsize=(9,5),sharex=True,sharey=False)\n",
    "fig.delaxes(ax[-1,-1])\n",
    "plt.setp(ax,xticks=[40,80],xticklabels=[40,80])\n",
    "ax = ax.reshape(-1)\n",
    "for ii,tag in enumerate(common_dict): \n",
    "    if genes_common[ii] != 'N/A':\n",
    "        ax[ii].set_title(genes_common[ii],fontsize=22)\n",
    "    else:\n",
    "        ax[ii].set_title(tags_common[ii],fontsize=22)\n",
    "    ax[ii].plot(corr_tspan,np.mean(data_fc[common_dict[tag][0]],axis=1))\n",
    "fig.text(0.55, -0.01, 'Time after malathion induction (minutes)', ha='center',fontsize=24);\n",
    "fig.text(-0.01, 0.5, 'Background subtracted TPM', va='center', rotation='vertical',fontsize=24);\n",
    "plt.tight_layout()\n",
    "# plt.savefig(figDir + 'supplementalFigs/dges_tpm.pdf',dpi=300,transparent=True,bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(2,3,figsize=(9,5),sharex=True,sharey=False)\n",
    "fig.delaxes(ax[-1,-1])\n",
    "plt.setp(ax,xticks=[40,80],xticklabels=[40,80])\n",
    "ax = ax.reshape(-1)\n",
    "for ii,tag in enumerate(common_dict): \n",
    "    if genes_common[ii] != 'N/A':\n",
    "        ax[ii].set_title(genes_common[ii],fontsize=22)\n",
    "    else:\n",
    "        ax[ii].set_title(tags_common[ii],fontsize=22)\n",
    "        \n",
    "        \n",
    "    pBS = data_t[common_dict[tag][1]] - data_c[common_dict[tag][1]]\n",
    "#     pBS = savgol( np.mean(data_t[common_dict[tag][1]],axis=1) - np.mean(data_c[common_dict[tag][1]],axis=1) ,5,2,axis=0)\n",
    "    mBS = np.mean(pBS,axis=1)\n",
    "    sBS = np.std(pBS,axis=1)    \n",
    "#     sBS =  np.sqrt(np.std(data_t[common_dict[tag][1]],axis=1)**2 + np.std(data_c[common_dict[tag][1]],axis=1)**2) \n",
    "#     ax[ii].plot(corr_tspan,pBS)\n",
    "    ax[ii].errorbar(corr_tspan,mBS,yerr=sBS,fmt='o--',color='k',lw=2,\n",
    "                    ms=9,mec='k',mew=1.5,capsize=3,elinewidth=2,ecolor='k') # mfc=cmap(ii)\n",
    "#     ax[ii].errorbar(corr_tspan,np.mean(data_t[common_dict[tag][1]],axis=1) - np.mean(data_c[common_dict[tag][1]],axis=1) -\n",
    "#                         (np.std(data_t[common_dict[tag][1]],axis=1) + np.std(data_c[common_dict[tag][1]],axis=1)), \n",
    "#                                    np.mean(data_t[common_dict[tag][1]],axis=1) - np.mean(data_c[common_dict[tag][1]],axis=1) + \n",
    "#                         (np.std(data_t[common_dict[tag][1]],axis=1) + np.std(data_c[common_dict[tag][1]],axis=1)),\n",
    "#                        color='tab:blue',alpha=0.2)\n",
    "    ax[ii].set_xlabel(f'Rank: {100*pr_deseq2[ii]:.2f}',fontsize=20,labelpad=2)\n",
    "\n",
    "fig.text(0.55, -0.01, 'Time after malathion induction (minutes)', ha='center',fontsize=24);\n",
    "fig.text(-0.01, 0.5, 'Background subtracted TPM', va='center', rotation='vertical',fontsize=24);\n",
    "plt.tight_layout()\n",
    "# plt.savefig(figDir + 'supplementalFigs/dges_tpm.pdf',dpi=300,transparent=True,bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# need to go back to txID to get logfoldchanges\n",
    "[de_corrected_tags[ii] for ii in [0,4,7,8,-2]], [deseq_df[deseq_df.padj < 0.05].tx_id.tolist()[ii] for ii in [0,4,7,8,-2]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "log2foldchanges = deseq_df[deseq_df.padj < 0.05].reset_index(drop=True).loc[[0,4,7,8,10]].log2FoldChange.tolist()\n",
    "log2foldchanges_se = deseq_df[deseq_df.padj < 0.05].reset_index(drop=True).loc[[0,4,7,8,10]].lfcSE.tolist()\n",
    "obsmaxranks = [np.argsort(np.abs(W[0])).tolist().index(inds_in_common_corrected[ii])/624 for ii in range(len(inds_in_common_corrected))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "[2**lfc for lfc in log2foldchanges]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Comparing the 180 differently expressed genes to the top 180 observable genes, we see an set intersection with size 30"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rank de_genes by their log2foldchange\n",
    "\n",
    "# first average values of duplicated tx_ids (differentially expressed in multiple time points)\n",
    "from collections import Counter\n",
    "dups = [k for (k,v) in Counter(deseq_df.tx_id.tolist()).items() if v > 1]\n",
    "\n",
    "log2FoldChanges = {}\n",
    "for tx in deseq_ids:\n",
    "    if tx in dups: \n",
    "        log2FoldChanges[tx] = deseq_df[deseq_df.tx_id.isin([tx])].log2FoldChange.mean()\n",
    "    else: \n",
    "        log2FoldChanges[tx] = list(deseq_df[deseq_df.tx_id.isin([tx])].log2FoldChange)[0]\n",
    "        \n",
    "log2FoldChanges = pd.DataFrame(log2FoldChanges.items(),columns=['tx_id','log2FoldChange'])\n",
    "\n",
    "# log2FoldChanges.tx_id to .tag\n",
    "de_genes, de_tags, _ = getRecords(cdsFasta_path,genbank_path,log2FoldChanges.tx_id.tolist())\n",
    "log2FoldChanges.tx_id = de_tags\n",
    "log2FoldChanges['gene'] = de_genes\n",
    "log2FoldChanges.columns = ['tag','log2FoldChange','gene']\n",
    "log2FoldChanges.sort_values(by='log2FoldChange',key=abs)\n",
    "log2FoldChanges['de_rank'] = list(range(len(log2FoldChanges)))\n",
    "# keep only the fold changes for genes that are also modeled with observability\n",
    "keep_tags = list(set(log2FoldChanges.tag.tolist()).intersection(locus_tags_keep))\n",
    "log2FoldChanges = log2FoldChanges[log2FoldChanges.tag.isin(keep_tags)]\n",
    "log2FoldChanges = log2FoldChanges.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "de_gene_tag = []\n",
    "for ii, gene in enumerate(log2FoldChanges.gene.tolist()): \n",
    "    if gene == 'N/A': \n",
    "        de_gene_tag.append(log2FoldChanges.tag.tolist()[ii])\n",
    "    else: \n",
    "        de_gene_tag.append(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_df = pd.DataFrame({sorted_tags[ii]: Wsorted[0,ii] for ii in range(len(sorted_tags))}.items(),columns=['tag','obs'])\n",
    "obs_df['obs_rank'] = list(range(len(obs_df)))\n",
    "obs_df = obs_df[obs_df.tag.isin(keep_tags)]\n",
    "obs_df = obs_df.set_index('tag')\n",
    "obs_df = obs_df.reindex(index=log2FoldChanges['tag'])\n",
    "obs_df = obs_df.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sp.stats.spearmanr(obs_df.obs_rank,log2FoldChanges.de_rank)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(\n",
    "    ((obs_df.obs_rank - min(obs_df.obs_rank)) / (max(obs_df.obs_rank) - min(obs_df.obs_rank))\n",
    "    ,(log2FoldChanges.de_rank-min(log2FoldChanges.de_rank))/\n",
    "     (max(log2FoldChanges.de_rank)-min(log2FoldChanges.de_rank)) ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "savefig = False\n",
    "\n",
    "plt.figure(figsize=(17,2))\n",
    "sns.heatmap(pd.DataFrame(\n",
    "    ((obs_df.obs_rank - min(obs_df.obs_rank)) / (max(obs_df.obs_rank) - min(obs_df.obs_rank))\n",
    "    ,(log2FoldChanges.de_rank-min(log2FoldChanges.de_rank))/\n",
    "     (max(log2FoldChanges.de_rank)-min(log2FoldChanges.de_rank)) )),\n",
    "     cmap='viridis',yticklabels=['observability rank','DGE rank'], xticklabels=de_gene_tag)\n",
    "plt.xticks(list(range(len(obs_df))),de_gene_tag,fontsize=13)\n",
    "# plt.tight_layout()\n",
    "if savefig: \n",
    "    plt.savefig(figDir + 'deseq_comparisons/methods_rank_comparison.pdf',bbox_inches='tight',transparent=True,dpi=300);\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# how many of the differentially expressed genes are in the actual library and what is their rank? \n",
    "log2FoldChanges[log2FoldChanges.tag.isin(df_up_new.locus_tag.tolist())]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### subfigures for sensor placement results "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# compare the eigenvalues of the three Gram matrices\n",
    "savefig=False\n",
    "# plt.style.use('dark_background')\n",
    "# dark_colors = plt.rcParams['axes.prop_cycle'].by_key()['color']  \n",
    "fig, ax = plt.subplots(1,1,figsize=(6,5))\n",
    "ax.plot(np.linspace(1,len(D),len(D)),np.real(D),'o',mfc='None',ms=8,mew=1.75,label=r'$G$')\n",
    "ax.plot(np.linspace(1,len(Dt),len(Dt)),np.real(Dt),'s',mfc='None',ms=8,mew=1.75,label=r'$\\tilde{G}$')\n",
    "# ax.plot(np.linspace(1,len(D),len(D)),np.real(Dt_recon),'v',mfc='None',ms=8,mew=1.75,label=r'$U_r\\tilde{G}U_r^\\top$')\n",
    "ax.set_yscale('log')\n",
    "ax.set_xscale('log')\n",
    "# ax.set_xticks([])\n",
    "ax.set_xlabel('gene')\n",
    "for side in ax.spines.keys():\n",
    "    ax.spines[side].set_linewidth(1.5)\n",
    "    ax.xaxis.set_tick_params(which='both', size=5, width=1.5, direction='in')\n",
    "    ax.yaxis.set_tick_params(which='both', size=5, width=1.5, direction='in')\n",
    "ax.set_ylabel(r'$\\lambda$',fontsize=30)\n",
    "plt.legend(frameon=False,ncol=1)\n",
    "if savefig: \n",
    "    plt.savefig(figDir + 'supplementalFigs/gram_eigvals.pdf',dpi=300,transparent=True,bbox_inches='tight')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compare top eigenvector of the three Gram matrices\n",
    "savefig=False\n",
    "# plt.style.use('dark_background')\n",
    "# dark_colors = plt.rcParams['axes.prop_cycle'].by_key()['color']  \n",
    "fig, ax = plt.subplots(1,1,figsize=(6,5))\n",
    "these_inds = np.argsort(V[:,0])\n",
    "ax.plot(V[:,0][these_inds],label=r'$G$')\n",
    "# ax.plot(np.linspace(0,623,10),np.sort(Vt[:,0]),label=r'$\\tilde{G}$')\n",
    "ax.plot((np.real(Vt_recon[:,0][these_inds]))[::-1],lw=2,label=r'$\\tilde{G}$') # actually eigenvec of projected up \\tilde{G} but to avoid confusion using \\tilde{G}\n",
    "# ax.set_xticks([])\n",
    "ax.set_xlabel('gene')\n",
    "for side in ax.spines.keys():\n",
    "    ax.spines[side].set_linewidth(1.5)\n",
    "    ax.xaxis.set_tick_params(which='both', size=5, width=1.5, direction='in')\n",
    "    ax.yaxis.set_tick_params(which='both', size=5, width=1.5, direction='in')\n",
    "ax.set_ylabel(r'$\\mathbf{w}_1$',fontsize=30)\n",
    "plt.legend(frameon=False,ncol=1)\n",
    "if savefig: \n",
    "    plt.savefig(figDir + 'supplementalFigs/gram_eigvec.pdf',dpi=300,transparent=True,bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# plot genes which contribute most and least to observability \n",
    "savefig = False\n",
    "ngenes = 20\n",
    "fig, ax = plt.subplots(2,1,figsize=(7,9.5),sharey=True) # 7,11\n",
    "for cnt,ii in enumerate(range(len(A)-1,len(A)-ngenes+1,-1)): \n",
    "    ax[0].set_title(str(ngenes)+' genes with highest \\n contribution to observability',fontsize=24)\n",
    "    mu = np.mean(data_fc_norm[sorted_inds_n[ii]],axis=1)\n",
    "    std = np.std(data_fc_norm[sorted_inds_n[ii]],axis=1)\n",
    "    ax[0].plot(corr_tspan,mu,'-',c='k',mfc='white',ms=2,mew=5,lw=0.75,zorder=1)\n",
    "    ax[0].fill_between(corr_tspan,mu-std,mu+std,color='k',alpha=0.05,zorder=0)\n",
    "ax[0].plot([10,90],[0,0],color='r',linestyle='-.',lw=2,zorder=1)\n",
    "# ax[0].set_ylabel('Fold change',fontsize=24)\n",
    "# plt.xlabel('Time after malathion induction (minutes)')\n",
    "# plt.savefig('/Users/aqib/desktop/top_genes.pdf',dpi=300,transparent=True,bbox_inches='tight')\n",
    "\n",
    "for cnt,ii in enumerate(range(ngenes)): \n",
    "    ax[1].set_title(str(ngenes)+' genes with lowest \\n contribution to observability',fontsize=24)\n",
    "    mu = np.mean(data_fc_norm[sorted_inds_n[ii]],axis=1)\n",
    "    std = np.std(data_fc_norm[sorted_inds_n[ii]],axis=1)\n",
    "    ax[1].plot(corr_tspan,mu,'-',c='k',mfc='white',ms=2,mew=5,lw=0.75,zorder=1)\n",
    "    ax[1].fill_between(corr_tspan,mu-std,mu+std,color='k',alpha=0.025,zorder=0)\n",
    "ax[1].fill_between([],[],[],color='k',alpha=0.2,label=r'$\\pm 1$ s.d.')\n",
    "ax[1].plot([10,90],[0,0],color='r',linestyle='-.',lw=2,zorder=1)\n",
    "# ax[1].set_ylabel('Fold change',fontsize=24)\n",
    "ax[1].set_xlabel('Time after malathion induction (minutes)',fontsize=26)\n",
    "ax[1].legend(frameon=False)\n",
    "fig.text(-0.05, 0.53, 'Standardized Fold Change', va='center', rotation='vertical',fontsize=29);\n",
    "for this_ax in ax.reshape(-1):\n",
    "    for side in this_ax.spines.keys():\n",
    "        this_ax.spines[side].set_linewidth(1.5)\n",
    "        this_ax.xaxis.set_tick_params(which='both', size=5, width=1.5, direction='in')\n",
    "        this_ax.yaxis.set_tick_params(which='both', size=5, width=1.5, direction='in')\n",
    "        \n",
    "plt.tight_layout()\n",
    "if savefig: \n",
    "    plt.savefig(figDir+'fig3/top_and_bottom_genes.pdf',dpi=300,transparent=True,bbox_inches='tight')\n",
    "plt.show()    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### DMD on each condition separately"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A_c,Atilde_c,data_fc_norm_red_c,U_r_c,cd_c,eVals_r_c,eVecs_r_c,Phi_c,b_r0_c,b_r1_c = \\\n",
    "        dmd(data_c[keepers],rank_reduce=True,r=10,trim=False,trimThresh=2.5e-3) \n",
    "A_t,Atilde_t,data_fc_norm_red_t,U_r_t,cd_t,eVals_r_t,eVecs_r_t,Phi_t,b_r0_t,b_r1_t = \\\n",
    "        dmd(data_t[keepers],rank_reduce=True,r=10,trim=False,trimThresh=2.5e-3) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Simulated systems and observability maximization"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1) Stable system, fully-connected, constant A "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "savefig = False\n",
    "\n",
    "np.random.seed(40)\n",
    "\n",
    "amin=1 # 0.001\n",
    "amax=1.5 # 0.1\n",
    "athresh=10\n",
    "nics=10\n",
    "x0mean=0.1\n",
    "x0std=0.1\n",
    "nblocks_list = [1,5,10,25,50]\n",
    "nvarsblock_list = [50,10,5,2,1]\n",
    "this_nT = 2\n",
    "\n",
    "fig, ax = plt.subplots(3,5,figsize=(12,6.5),sharex='row',sharey='row',gridspec_kw={'height_ratios':[1.2,1,1]})\n",
    "# plt.setp(ax,xticks=[0,50,100],xticklabels=[0,50,100])\n",
    "\n",
    "for ii in range(len(nblocks_list)):\n",
    "    nblocks = nblocks_list[ii]\n",
    "    nvarsblock = nvarsblock_list[ii]\n",
    "    Asim , x0sim = gen_block_lin_sys(N_blocks=nblocks,N_vars_per_block=nvarsblock,Ai_min=amin,Ai_max=amax,N_initial_conditions=nics,x0_mean=x0mean,x0_std=x0std)\n",
    "    Gsim = gram_matrix(Asim,x0sim,nT=this_nT,reduced=False)\n",
    "    Dsim, Vsim = sp.linalg.eig(Gsim,left=False,right=True)\n",
    "    Wsim = Vsim[:,0:1].T[0]\n",
    "#     Wsim = (Wsim - np.min(Wsim)) / (np.max(Wsim) - np.min(Wsim))\n",
    "        \n",
    "    sns.heatmap(Asim,ax=ax[0,ii],cbar=False,cmap='viridis')\n",
    "    ax[0,ii].set_xticklabels(labels=[]); ax[0,ii].set_yticklabels(labels=[])\n",
    "    ax[1,ii].plot(Wsim)\n",
    "    ax[1,ii].plot([0,50],[0,0],'--k',lw=1.5,zorder=0)\n",
    "    ax[2,ii].plot(np.sort(np.abs(Wsim)))\n",
    "    ax[2,ii].plot([0,50],[0,0],'--k',lw=1.5,zorder=0)\n",
    "\n",
    "ax[-1,2].set_xlabel('state variable index')    \n",
    "ax[0,0].set_ylabel('State- \\n transitions'); ax[1,0].set_ylabel(r'$\\mathbf{w}$'); ax[2,0].set_ylabel(r'$\\mathbf{w}$ (scaled)' + '\\n sorted'); \n",
    "plt.suptitle(f'T = {this_nT}',fontsize=21)\n",
    "plt.tight_layout()\n",
    "if savefig: \n",
    "    plt.savefig(figDir+'supplementalFigs/obs_max_expos_T2.pdf',dpi=300,transparent=True,bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "savefig = False\n",
    "\n",
    "np.random.seed(40)\n",
    "\n",
    "amin=1 # 0.001\n",
    "amax=1.5 # 0.1\n",
    "athresh=10\n",
    "nics=10\n",
    "x0mean=0.1\n",
    "x0std=0.1\n",
    "nblocks_list = [1,5,10,25,50]\n",
    "nvarsblock_list = [50,10,5,2,1]\n",
    "this_nT = 5\n",
    "\n",
    "fig, ax = plt.subplots(3,5,figsize=(12,6.5),sharex='row',sharey='row',gridspec_kw={'height_ratios':[1.2,1,1]})\n",
    "# plt.setp(ax,xticks=[0,50,100],xticklabels=[0,50,100])\n",
    "\n",
    "for ii in range(len(nblocks_list)):\n",
    "    nblocks = nblocks_list[ii]\n",
    "    nvarsblock = nvarsblock_list[ii]\n",
    "    Asim , x0sim = gen_block_lin_sys(N_blocks=nblocks,N_vars_per_block=nvarsblock,Ai_min=amin,Ai_max=amax,N_initial_conditions=nics,x0_mean=x0mean,x0_std=x0std)\n",
    "    Gsim = gram_matrix(Asim,x0sim,nT=this_nT,reduced=False)\n",
    "    Dsim, Vsim = sp.linalg.eig(Gsim,left=False,right=True)\n",
    "    Wsim = Vsim[:,0:1].T[0]\n",
    "#     Wsim = (Wsim - np.min(Wsim)) / (np.max(Wsim) - np.min(Wsim))\n",
    "        \n",
    "    sns.heatmap(Asim,ax=ax[0,ii],cbar=False,cmap='viridis')\n",
    "    ax[0,ii].set_xticklabels(labels=[]); ax[0,ii].set_yticklabels(labels=[])\n",
    "    ax[1,ii].plot(Wsim)\n",
    "    ax[1,ii].plot([0,50],[0,0],'--k',lw=1.5,zorder=0)\n",
    "    ax[2,ii].plot(np.sort(np.abs(Wsim)))\n",
    "    ax[2,ii].plot([0,50],[0,0],'--k',lw=1.5,zorder=0)\n",
    "\n",
    "ax[-1,2].set_xlabel('state variable index')    \n",
    "ax[0,0].set_ylabel('State- \\n transitions'); ax[1,0].set_ylabel(r'$\\mathbf{w}$'); ax[2,0].set_ylabel(r'$\\mathbf{w}$ (scaled)' + '\\n sorted'); \n",
    "plt.suptitle(f'T = {this_nT}',fontsize=21)\n",
    "plt.tight_layout()\n",
    "if savefig: \n",
    "    plt.savefig(figDir+'supplementalFigs/obs_max_expos_T5.pdf',dpi=300,transparent=True,bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "savfig = False\n",
    "\n",
    "np.random.seed(40)\n",
    "\n",
    "amin=1 # 0.001\n",
    "amax=1.5 # 0.1\n",
    "athresh=10\n",
    "nics=10\n",
    "x0mean=0.1\n",
    "x0std=0.1\n",
    "nblocks_list = [1,5,10,25,50]\n",
    "nvarsblock_list = [50,10,5,2,1]\n",
    "this_nT = 2\n",
    "\n",
    "fig, ax = plt.subplots(3,5,figsize=(12,6.5),sharex='row',sharey='row',gridspec_kw={'height_ratios':[1.2,1,1]})\n",
    "# plt.setp(ax,xticks=[0,50,100],xticklabels=[0,50,100])\n",
    "\n",
    "for ii in range(len(nblocks_list)):\n",
    "    nblocks = nblocks_list[ii]\n",
    "    nvarsblock = nvarsblock_list[ii]\n",
    "    Asim , x0sim = gen_block_lin_sys(N_blocks=nblocks,N_vars_per_block=nvarsblock,Ai_min=amin,Ai_max=amax,N_initial_conditions=nics,x0_mean=x0mean,x0_std=x0std)\n",
    "    Asim[-20:,-20:] = 0.0\n",
    "    Gsim = gram_matrix(Asim,x0sim,nT=this_nT,reduced=False)\n",
    "    Dsim, Vsim = sp.linalg.eig(Gsim,left=False,right=True)\n",
    "    Wsim = Vsim[:,0:1].T[0]\n",
    "#     Wsim = (Wsim - np.min(Wsim)) / (np.max(Wsim) - np.min(Wsim))\n",
    "        \n",
    "    sns.heatmap(Asim,ax=ax[0,ii],cbar=False,cmap='viridis')\n",
    "    ax[0,ii].set_xticklabels(labels=[]); ax[0,ii].set_yticklabels(labels=[])\n",
    "    ax[1,ii].plot(Wsim)\n",
    "    ax[1,ii].plot([0,50],[0,0],'--k',lw=1.5,zorder=0)\n",
    "    ax[2,ii].plot(np.sort(np.abs(Wsim)))\n",
    "    ax[2,ii].plot([0,50],[0,0],'--k',lw=1.5,zorder=0)\n",
    "\n",
    "ax[-1,2].set_xlabel('state variable index')    \n",
    "ax[0,0].set_ylabel('State- \\n transitions'); ax[1,0].set_ylabel(r'$\\mathbf{w}$'); ax[2,0].set_ylabel(r'$\\mathbf{w}$ (scaled)' + '\\n sorted'); \n",
    "plt.suptitle(f'T = {this_nT}',fontsize=21)\n",
    "plt.tight_layout()\n",
    "if savefig: \n",
    "    plt.savefig(figDir+'supplementalFigs/obs_max_expos_T2_zeros.pdf',dpi=300,transparent=True,bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "savefig = False\n",
    "\n",
    "np.random.seed(40)\n",
    "\n",
    "amin=1 # 0.001\n",
    "amax=1.5 # 0.1\n",
    "athresh=10\n",
    "nics=10\n",
    "x0mean=0.1\n",
    "x0std=0.1\n",
    "nblocks_list = [1,5,10,25,50]\n",
    "nvarsblock_list = [50,10,5,2,1]\n",
    "this_nT = 5\n",
    "\n",
    "fig, ax = plt.subplots(3,5,figsize=(12,6.5),sharex='row',sharey='row',gridspec_kw={'height_ratios':[1.2,1,1]})\n",
    "# plt.setp(ax,xticks=[0,50,100],xticklabels=[0,50,100])\n",
    "\n",
    "for ii in range(len(nblocks_list)):\n",
    "    nblocks = nblocks_list[ii]\n",
    "    nvarsblock = nvarsblock_list[ii]\n",
    "    Asim , x0sim = gen_block_lin_sys(N_blocks=nblocks,N_vars_per_block=nvarsblock,Ai_min=amin,Ai_max=amax,N_initial_conditions=nics,x0_mean=x0mean,x0_std=x0std)\n",
    "    Asim[-20:,-20:] = 0.0\n",
    "    Gsim = gram_matrix(Asim,x0sim,nT=this_nT,reduced=False)\n",
    "    Dsim, Vsim = sp.linalg.eig(Gsim,left=False,right=True)\n",
    "    Wsim = Vsim[:,0:1].T[0]\n",
    "#     Wsim = (Wsim - np.min(Wsim)) / (np.max(Wsim) - np.min(Wsim))\n",
    "        \n",
    "    sns.heatmap(Asim,ax=ax[0,ii],cbar=False,cmap='viridis')\n",
    "    ax[0,ii].set_xticklabels(labels=[]); ax[0,ii].set_yticklabels(labels=[])\n",
    "    ax[1,ii].plot(Wsim)\n",
    "    ax[1,ii].plot([0,50],[0,0],'--k',lw=1.5,zorder=0)\n",
    "    ax[2,ii].plot(np.sort(np.abs(Wsim)))\n",
    "    ax[2,ii].plot([0,50],[0,0],'--k',lw=1.5,zorder=0)\n",
    "\n",
    "\n",
    "ax[-1,2].set_xlabel('state variable index')\n",
    "ax[0,0].set_ylabel('State- \\n transitions'); ax[1,0].set_ylabel(r'$\\mathbf{w}$'); ax[2,0].set_ylabel(r'$\\mathbf{w}$ (scaled)' + '\\n sorted'); \n",
    "plt.suptitle(f'T = {this_nT}',fontsize=21)\n",
    "plt.tight_layout()\n",
    "if savefig: \n",
    "    plt.savefig(figDir+'supplementalFigs/obs_max_expos_T5_zeros.pdf',dpi=300,transparent=True,bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What pathways are the top 20 genes involved in?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# protein name, molecular function, biological process\n",
    "\n",
    "pd.DataFrame({'locus_tag': [locus_tags_keep[x] for x in sorted_inds_n[::-1][:20]],\n",
    " 'gene': [genes_keep[x] for x in sorted_inds_n[::-1][:20]],\n",
    " 'protein_name':['50S ribosomal protein L34',\n",
    "                'Putative outer membrane protein',\n",
    "                'ATP-dependent Clp protease proteolytic subunit',\n",
    "                'Translational regulator CsrA',\n",
    "                'ATP synthase subunit a',\n",
    "                '30S ribosomal protein S11',\n",
    "                '50S ribosomal protein L6',\n",
    "                '50S ribosomal protein L10',\n",
    "                'ATP synthase subunit b',\n",
    "                'Coenzyme PQQ synthesis protein A',\n",
    "                'Ribose-phosphate pyrophosphokinase',\n",
    "                'Probable Fe(2+)-trafficking protein',\n",
    "                '30S ribosomal protein S7',\n",
    "                'Arginine deiminase',\n",
    "                '50S ribosomal protein L16',\n",
    "                '30S ribosomal protein S13',\n",
    "                'Phenylalanine--tRNA ligase alpha subunit',\n",
    "                '30S ribosomal protein S2',\n",
    "                '50S ribosomal protein L3',\n",
    "                'DUF1161 domain-containing protein'],\n",
    "'molecular function':['structural constituent of ribosome',\n",
    "                      'N/A',\n",
    "                      'ATP-dependent peptidase activity; serine-type endopeptidase activity',\n",
    "                      \"mRNA 5'-UTR binding\",\"proton-transporting ATP synthase activity, rotational mechanism\",\n",
    "                      'structural constituent of ribosome',\n",
    "                      'structural constituent of ribosome',\n",
    "                      'structural constituent of ribosome',\n",
    "                      'hydrolase activity; proton-transporting ATP synthase activity, rotational mechanism',\n",
    "                      'N/A',\n",
    "                      'ATP binding; kinase activity; magnesium ion binding; ribose phosphate diphosphokinase activiity',\n",
    "                      'N/A',\n",
    "                      'rRNA binding; structural constituent of ribosome; tRNA binding',\n",
    "                      'arginine deiminase activity',\n",
    "                      'rRNA binding; structural constituent of ribosome; tRNA binding',\n",
    "                      'rRNA binding; structural constituent of ribosome; tRNA binding',\n",
    "                      'ATP binding; magnesium ion binding; phenylalanine-tRNA ligase activity; tRNA binding',\n",
    "                      'structural constituent of ribosome',\n",
    "                      'rRNA binding; structural constituent of ribosome',\n",
    "                      'N/A'],\n",
    "'biological process':['translation',\n",
    "                     'N/A',\n",
    "                     'proteolysis',\n",
    "                     'regulation of translational initation and carbohydrate metabolic process',\n",
    "                     'N/A',\n",
    "                     'translation',\n",
    "                     'translation',\n",
    "                     'translation',\n",
    "                     'N/A',\n",
    "                     'pyrroloquinoline quinone biosynthetic process',\n",
    "                     '5-phosphoribose 1-diphosphate biosynthetic process; nucleotide biosynthetic process; phosphorylation; ribonucleoside monophosphate biosynthetic process',\n",
    "                     'N/A',\n",
    "                     'translation',\n",
    "                     'arginine catabolic process to  ornithine',\n",
    "                     'translation',\n",
    "                     'translation',\n",
    "                     'phenylalanyl-tRNA aminoacylation',\n",
    "                     'translation',\n",
    "                     'translation',\n",
    "                     'N/A'] })"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "2.7.undefined"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
